// @wordpress/editor/build/components/document-bar/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DocumentBar;
var _clsx = _interopRequireDefault(require("clsx"));
var _i18n = require("@wordpress/i18n");
var _data = require("@wordpress/data");
var _components = require("@wordpress/components");
var _blockEditor = require("@wordpress/block-editor");
var _icons = require("@wordpress/icons");
var _keycodes = require("@wordpress/keycodes");
var _coreData = require("@wordpress/core-data");
var _commands = require("@wordpress/commands");
var _element = require("@wordpress/element");
var _compose = require("@wordpress/compose");
var _htmlEntities = require("@wordpress/html-entities");
var _constants = require("../../store/constants");
var _store = require("../../store");
var _lockUnlock = require("../../lock-unlock");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const TYPE_LABELS = {
  // translators: 1: Pattern title.
  wp_pattern: (0, _i18n.__)('Editing pattern: %s'),
  // translators: 1: Navigation menu title.
  wp_navigation: (0, _i18n.__)('Editing navigation menu: %s'),
  // translators: 1: Template title.
  wp_template: (0, _i18n.__)('Editing template: %s'),
  // translators: 1: Template part title.
  wp_template_part: (0, _i18n.__)('Editing template part: %s')
};
const MotionButton = (0, _components.__unstableMotion)(_components.Button);

/**
 * This component renders a navigation bar at the top of the editor. It displays the title of the current document,
 * a back button (if applicable), and a command center button. It also handles different states of the document,
 * such as "not found" or "unsynced".
 *
 * @example
 * ```jsx
 * <DocumentBar />
 * ```
 * @param {Object}                                   props       The component props.
 * @param {string}                                   props.title A title for the document, defaulting to the document or
 *                                                               template title currently being edited.
 * @param {import("@wordpress/components").IconType} props.icon  An icon for the document, defaulting to an icon for document
 *                                                               or template currently being edited.
 *
 * @return {JSX.Element} The rendered DocumentBar component.
 */
function DocumentBar(props) {
  const {
    postType,
    documentTitle,
    isNotFound,
    isUnsyncedPattern,
    templateIcon,
    templateTitle,
    onNavigateToPreviousEntityRecord
  } = (0, _data.useSelect)(select => {
    const {
      getCurrentPostType,
      getCurrentPostId,
      getEditorSettings,
      __experimentalGetTemplateInfo: getTemplateInfo
    } = select(_store.store);
    const {
      getEditedEntityRecord,
      isResolving: isResolvingSelector
    } = select(_coreData.store);
    const _postType = getCurrentPostType();
    const _postId = getCurrentPostId();
    const _document = getEditedEntityRecord('postType', _postType, _postId);
    const _templateInfo = getTemplateInfo(_document);
    return {
      postType: _postType,
      documentTitle: _document.title,
      isNotFound: !_document && !isResolvingSelector('getEditedEntityRecord', 'postType', _postType, _postId),
      isUnsyncedPattern: _document?.wp_pattern_sync_status === 'unsynced',
      templateIcon: (0, _lockUnlock.unlock)(select(_store.store)).getPostIcon(_postType, {
        area: _document?.area
      }),
      templateTitle: _templateInfo.title,
      onNavigateToPreviousEntityRecord: getEditorSettings().onNavigateToPreviousEntityRecord
    };
  }, []);
  const {
    open: openCommandCenter
  } = (0, _data.useDispatch)(_commands.store);
  const isReducedMotion = (0, _compose.useReducedMotion)();
  const isTemplate = _constants.TEMPLATE_POST_TYPES.includes(postType);
  const isGlobalEntity = _constants.GLOBAL_POST_TYPES.includes(postType);
  const hasBackButton = !!onNavigateToPreviousEntityRecord;
  const entityTitle = isTemplate ? templateTitle : documentTitle;
  const title = props.title || entityTitle;
  const icon = props.icon || templateIcon;
  const mounted = (0, _element.useRef)(false);
  (0, _element.useEffect)(() => {
    mounted.current = true;
  }, []);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    className: (0, _clsx.default)('editor-document-bar', {
      'has-back-button': hasBackButton,
      'is-global': isGlobalEntity && !isUnsyncedPattern
    }),
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__unstableAnimatePresence, {
      children: hasBackButton && /*#__PURE__*/(0, _jsxRuntime.jsx)(MotionButton, {
        className: "editor-document-bar__back",
        icon: (0, _i18n.isRTL)() ? _icons.chevronRightSmall : _icons.chevronLeftSmall,
        onClick: event => {
          event.stopPropagation();
          onNavigateToPreviousEntityRecord();
        },
        size: "compact",
        initial: mounted.current ? {
          opacity: 0,
          transform: 'translateX(15%)'
        } : false // Don't show entry animation when DocumentBar mounts.
        ,
        animate: {
          opacity: 1,
          transform: 'translateX(0%)'
        },
        exit: {
          opacity: 0,
          transform: 'translateX(15%)'
        },
        transition: isReducedMotion ? {
          duration: 0
        } : undefined,
        children: (0, _i18n.__)('Back')
      })
    }), isNotFound ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalText, {
      children: (0, _i18n.__)('Document not found')
    }) : /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.Button, {
      className: "editor-document-bar__command",
      onClick: () => openCommandCenter(),
      size: "compact",
      children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__unstableMotion.div, {
        className: "editor-document-bar__title"
        // Force entry animation when the back button is added or removed.
        ,

        initial: mounted.current ? {
          opacity: 0,
          transform: hasBackButton ? 'translateX(15%)' : 'translateX(-15%)'
        } : false // Don't show entry animation when DocumentBar mounts.
        ,
        animate: {
          opacity: 1,
          transform: 'translateX(0%)'
        },
        transition: isReducedMotion ? {
          duration: 0
        } : undefined,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.BlockIcon, {
          icon: icon
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalText, {
          size: "body",
          as: "h1",
          "aria-label": !props.title && TYPE_LABELS[postType] ?
          // eslint-disable-next-line @wordpress/valid-sprintf
          (0, _i18n.sprintf)(TYPE_LABELS[postType], title) : undefined,
          children: title ? (0, _htmlEntities.decodeEntities)(title) : (0, _i18n.__)('No title')
        })]
      }, hasBackButton), /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
        className: "editor-document-bar__shortcut",
        children: _keycodes.displayShortcut.primary('k')
      })]
    })]
  });
}
