// @ariakit/react-core/cjs/__chunks/55IP6UF7.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";






var _IHRK7HOTcjs = require('./IHRK7HOT.cjs');



var _BZ34A7Q7cjs = require('./BZ34A7Q7.cjs');



var _KOJWH5YCcjs = require('./KOJWH5YC.cjs');


var _AWEKOM65cjs = require('./AWEKOM65.cjs');

// src/menu/menu-context.tsx
var _react = require('react');
var menu = _AWEKOM65cjs.createStoreContext.call(void 0, 
  [_BZ34A7Q7cjs.CompositeContextProvider, _KOJWH5YCcjs.HovercardContextProvider],
  [_BZ34A7Q7cjs.CompositeScopedContextProvider, _KOJWH5YCcjs.HovercardScopedContextProvider]
);
var useMenuContext = menu.useContext;
var useMenuScopedContext = menu.useScopedContext;
var useMenuProviderContext = menu.useProviderContext;
var MenuContextProvider = menu.ContextProvider;
var MenuScopedContextProvider = menu.ScopedContextProvider;
var useMenuBarContext = _IHRK7HOTcjs.useMenubarContext;
var useMenuBarScopedContext = _IHRK7HOTcjs.useMenubarScopedContext;
var useMenuBarProviderContext = _IHRK7HOTcjs.useMenubarProviderContext;
var MenuBarContextProvider = _IHRK7HOTcjs.MenubarContextProvider;
var MenuBarScopedContextProvider = _IHRK7HOTcjs.MenubarScopedContextProvider;
var MenuItemCheckedContext = _react.createContext.call(void 0, 
  void 0
);













exports.useMenuContext = useMenuContext; exports.useMenuScopedContext = useMenuScopedContext; exports.useMenuProviderContext = useMenuProviderContext; exports.MenuContextProvider = MenuContextProvider; exports.MenuScopedContextProvider = MenuScopedContextProvider; exports.useMenuBarContext = useMenuBarContext; exports.useMenuBarScopedContext = useMenuBarScopedContext; exports.useMenuBarProviderContext = useMenuBarProviderContext; exports.MenuBarContextProvider = MenuBarContextProvider; exports.MenuBarScopedContextProvider = MenuBarScopedContextProvider; exports.MenuItemCheckedContext = MenuItemCheckedContext;
