// @ariakit/react-core/cjs/__chunks/BZ34A7Q7.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _55G3CDVQcjs = require('./55G3CDVQ.cjs');


var _AWEKOM65cjs = require('./AWEKOM65.cjs');

// src/composite/composite-context.tsx
var _react = require('react');
var ctx = _AWEKOM65cjs.createStoreContext.call(void 0, 
  [_55G3CDVQcjs.CollectionContextProvider],
  [_55G3CDVQcjs.CollectionScopedContextProvider]
);
var useCompositeContext = ctx.useContext;
var useCompositeScopedContext = ctx.useScopedContext;
var useCompositeProviderContext = ctx.useProviderContext;
var CompositeContextProvider = ctx.ContextProvider;
var CompositeScopedContextProvider = ctx.ScopedContextProvider;
var CompositeItemContext = _react.createContext.call(void 0, 
  void 0
);
var CompositeRowContext = _react.createContext.call(void 0, 
  void 0
);









exports.useCompositeContext = useCompositeContext; exports.useCompositeScopedContext = useCompositeScopedContext; exports.useCompositeProviderContext = useCompositeProviderContext; exports.CompositeContextProvider = CompositeContextProvider; exports.CompositeScopedContextProvider = CompositeScopedContextProvider; exports.CompositeItemContext = CompositeItemContext; exports.CompositeRowContext = CompositeRowContext;
