// @ariakit/react-core/cjs/__chunks/CQ3ILEKO.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _NRLE3XE7cjs = require('./NRLE3XE7.cjs');


var _AWEKOM65cjs = require('./AWEKOM65.cjs');

// src/dialog/dialog-context.tsx
var _react = require('react');
var ctx = _AWEKOM65cjs.createStoreContext.call(void 0, 
  [_NRLE3XE7cjs.DisclosureContextProvider],
  [_NRLE3XE7cjs.DisclosureScopedContextProvider]
);
var useDialogContext = ctx.useContext;
var useDialogScopedContext = ctx.useScopedContext;
var useDialogProviderContext = ctx.useProviderContext;
var DialogContextProvider = ctx.ContextProvider;
var DialogScopedContextProvider = ctx.ScopedContextProvider;
var DialogHeadingContext = _react.createContext.call(void 0, void 0);
var DialogDescriptionContext = _react.createContext.call(void 0, void 0);









exports.useDialogContext = useDialogContext; exports.useDialogScopedContext = useDialogScopedContext; exports.useDialogProviderContext = useDialogProviderContext; exports.DialogContextProvider = DialogContextProvider; exports.DialogScopedContextProvider = DialogScopedContextProvider; exports.DialogHeadingContext = DialogHeadingContext; exports.DialogDescriptionContext = DialogDescriptionContext;
