// @ariakit/react-core/cjs/__chunks/DPVDBBFO.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _35VA3E7Jcjs = require('./35VA3E7J.cjs');




var _AWEKOM65cjs = require('./AWEKOM65.cjs');



var _OUWHJLQHcjs = require('./OUWHJLQH.cjs');



var _3IVNQNIVcjs = require('./3IVNQNIV.cjs');

// src/button/button.tsx
var _dom = require('@ariakit/core/utils/dom');
var _react = require('react');
var TagName = "button";
var useButton = _AWEKOM65cjs.createHook.call(void 0, 
  function useButton2(props) {
    const ref = _react.useRef.call(void 0, null);
    const tagName = _OUWHJLQHcjs.useTagName.call(void 0, ref, TagName);
    const [isNativeButton, setIsNativeButton] = _react.useState.call(void 0, 
      () => !!tagName && _dom.isButton.call(void 0, { tagName, type: props.type })
    );
    _react.useEffect.call(void 0, () => {
      if (!ref.current) return;
      setIsNativeButton(_dom.isButton.call(void 0, ref.current));
    }, []);
    props = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {
      role: !isNativeButton && tagName !== "a" ? "button" : void 0
    }, props), {
      ref: _OUWHJLQHcjs.useMergeRefs.call(void 0, ref, props.ref)
    });
    props = _35VA3E7Jcjs.useCommand.call(void 0, props);
    return props;
  }
);
var Button = _AWEKOM65cjs.forwardRef.call(void 0, function Button2(props) {
  const htmlProps = useButton(props);
  return _AWEKOM65cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useButton = useButton; exports.Button = Button;
