// @ariakit/react-core/cjs/__chunks/I2B6AUM7.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _TILYAEREcjs = require('./TILYAERE.cjs');




var _AWEKOM65cjs = require('./AWEKOM65.cjs');



var _3IVNQNIVcjs = require('./3IVNQNIV.cjs');

// src/focus-trap/focus-trap.tsx
var TagName = "span";
var useFocusTrap = _AWEKOM65cjs.createHook.call(void 0, 
  function useFocusTrap2(props) {
    props = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {
      "data-focus-trap": "",
      tabIndex: 0,
      "aria-hidden": true
    }, props), {
      style: _3IVNQNIVcjs.__spreadValues.call(void 0, {
        // Prevents unintended scroll jumps.
        position: "fixed",
        top: 0,
        left: 0
      }, props.style)
    });
    props = _TILYAEREcjs.useVisuallyHidden.call(void 0, props);
    return props;
  }
);
var FocusTrap = _AWEKOM65cjs.forwardRef.call(void 0, function FocusTrap2(props) {
  const htmlProps = useFocusTrap(props);
  return _AWEKOM65cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useFocusTrap = useFocusTrap; exports.FocusTrap = FocusTrap;
