// @ariakit/react-core/cjs/__chunks/I52OTQQ2.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _BZ34A7Q7cjs = require('./BZ34A7Q7.cjs');



var _5XC2EBLVcjs = require('./5XC2EBLV.cjs');


var _AWEKOM65cjs = require('./AWEKOM65.cjs');

// src/select/select-context.tsx
var _react = require('react');
var ctx = _AWEKOM65cjs.createStoreContext.call(void 0, 
  [_5XC2EBLVcjs.PopoverContextProvider, _BZ34A7Q7cjs.CompositeContextProvider],
  [_5XC2EBLVcjs.PopoverScopedContextProvider, _BZ34A7Q7cjs.CompositeScopedContextProvider]
);
var useSelectContext = ctx.useContext;
var useSelectScopedContext = ctx.useScopedContext;
var useSelectProviderContext = ctx.useProviderContext;
var SelectContextProvider = ctx.ContextProvider;
var SelectScopedContextProvider = ctx.ScopedContextProvider;
var SelectItemCheckedContext = _react.createContext.call(void 0, false);
var SelectHeadingContext = _react.createContext.call(void 0, null);









exports.useSelectContext = useSelectContext; exports.useSelectScopedContext = useSelectScopedContext; exports.useSelectProviderContext = useSelectProviderContext; exports.SelectContextProvider = SelectContextProvider; exports.SelectScopedContextProvider = SelectScopedContextProvider; exports.SelectItemCheckedContext = SelectItemCheckedContext; exports.SelectHeadingContext = SelectHeadingContext;
