// @ariakit/react-core/cjs/__chunks/IHRK7HOT.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _BZ34A7Q7cjs = require('./BZ34A7Q7.cjs');


var _AWEKOM65cjs = require('./AWEKOM65.cjs');

// src/menubar/menubar-context.tsx
var _react = require('react');
var menubar = _AWEKOM65cjs.createStoreContext.call(void 0, 
  [_BZ34A7Q7cjs.CompositeContextProvider],
  [_BZ34A7Q7cjs.CompositeScopedContextProvider]
);
var useMenubarContext = menubar.useContext;
var useMenubarScopedContext = menubar.useScopedContext;
var useMenubarProviderContext = menubar.useProviderContext;
var MenubarContextProvider = menubar.ContextProvider;
var MenubarScopedContextProvider = menubar.ScopedContextProvider;
var MenuItemCheckedContext = _react.createContext.call(void 0, 
  void 0
);








exports.useMenubarContext = useMenubarContext; exports.useMenubarScopedContext = useMenubarScopedContext; exports.useMenubarProviderContext = useMenubarProviderContext; exports.MenubarContextProvider = MenubarContextProvider; exports.MenubarScopedContextProvider = MenubarScopedContextProvider; exports.MenuItemCheckedContext = MenuItemCheckedContext;
