// @ariakit/react-core/cjs/__chunks/L2VXOCPG.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _DPVDBBFOcjs = require('./DPVDBBFO.cjs');


var _CQ3ILEKOcjs = require('./CQ3ILEKO.cjs');




var _AWEKOM65cjs = require('./AWEKOM65.cjs');


var _OUWHJLQHcjs = require('./OUWHJLQH.cjs');




var _3IVNQNIVcjs = require('./3IVNQNIV.cjs');

// src/dialog/dialog-dismiss.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "button";
var useDialogDismiss = _AWEKOM65cjs.createHook.call(void 0, 
  function useDialogDismiss2(_a) {
    var _b = _a, { store } = _b, props = _3IVNQNIVcjs.__objRest.call(void 0, _b, ["store"]);
    const context = _CQ3ILEKOcjs.useDialogScopedContext.call(void 0, );
    store = store || context;
    const onClickProp = props.onClick;
    const onClick = _OUWHJLQHcjs.useEvent.call(void 0, (event) => {
      onClickProp == null ? void 0 : onClickProp(event);
      if (event.defaultPrevented) return;
      store == null ? void 0 : store.hide();
    });
    const children = _react.useMemo.call(void 0, 
      () => /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, 
        "svg",
        {
          "aria-label": "Dismiss popup",
          display: "block",
          fill: "none",
          stroke: "currentColor",
          strokeLinecap: "round",
          strokeLinejoin: "round",
          strokeWidth: 1.5,
          viewBox: "0 0 16 16",
          height: "1em",
          width: "1em",
          children: [
            /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", { x1: "4", y1: "4", x2: "12", y2: "12" }),
            /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", { x1: "4", y1: "12", x2: "12", y2: "4" })
          ]
        }
      ),
      []
    );
    props = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {
      "data-dialog-dismiss": "",
      children
    }, props), {
      onClick
    });
    props = _DPVDBBFOcjs.useButton.call(void 0, props);
    return props;
  }
);
var DialogDismiss = _AWEKOM65cjs.forwardRef.call(void 0, function DialogDismiss2(props) {
  const htmlProps = useDialogDismiss(props);
  return _AWEKOM65cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useDialogDismiss = useDialogDismiss; exports.DialogDismiss = DialogDismiss;
