// @ariakit/react-core/cjs/__chunks/LO2HIMYX.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _BZ34A7Q7cjs = require('./BZ34A7Q7.cjs');


var _AWEKOM65cjs = require('./AWEKOM65.cjs');

// src/tag/tag-context.tsx
var _react = require('react');
var TagValueContext = _react.createContext.call(void 0, null);
var TagRemoveIdContext = _react.createContext.call(void 0, 
  null
);
var ctx = _AWEKOM65cjs.createStoreContext.call(void 0, 
  [_BZ34A7Q7cjs.CompositeContextProvider],
  [_BZ34A7Q7cjs.CompositeScopedContextProvider]
);
var useTagContext = ctx.useContext;
var useTagScopedContext = ctx.useScopedContext;
var useTagProviderContext = ctx.useProviderContext;
var TagContextProvider = ctx.ContextProvider;
var TagScopedContextProvider = ctx.ScopedContextProvider;









exports.TagValueContext = TagValueContext; exports.TagRemoveIdContext = TagRemoveIdContext; exports.useTagContext = useTagContext; exports.useTagScopedContext = useTagScopedContext; exports.useTagProviderContext = useTagProviderContext; exports.TagContextProvider = TagContextProvider; exports.TagScopedContextProvider = TagScopedContextProvider;
