// @ariakit/react-core/cjs/__chunks/PYKELHKQ.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _55G3CDVQcjs = require('./55G3CDVQ.cjs');




var _AWEKOM65cjs = require('./AWEKOM65.cjs');



var _OUWHJLQHcjs = require('./OUWHJLQH.cjs');




var _3IVNQNIVcjs = require('./3IVNQNIV.cjs');

// src/collection/collection-item.tsx
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var TagName = "div";
var useCollectionItem = _AWEKOM65cjs.createHook.call(void 0, 
  function useCollectionItem2(_a) {
    var _b = _a, {
      store,
      shouldRegisterItem = true,
      getItem = _misc.identity,
      element: element
    } = _b, props = _3IVNQNIVcjs.__objRest.call(void 0, _b, [
      "store",
      "shouldRegisterItem",
      "getItem",
      // @ts-expect-error This prop may come from a collection renderer.
      "element"
    ]);
    const context = _55G3CDVQcjs.useCollectionContext.call(void 0, );
    store = store || context;
    const id = _OUWHJLQHcjs.useId.call(void 0, props.id);
    const ref = _react.useRef.call(void 0, element);
    _react.useEffect.call(void 0, () => {
      const element2 = ref.current;
      if (!id) return;
      if (!element2) return;
      if (!shouldRegisterItem) return;
      const item = getItem({ id, element: element2 });
      return store == null ? void 0 : store.renderItem(item);
    }, [id, shouldRegisterItem, getItem, store]);
    props = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, props), {
      ref: _OUWHJLQHcjs.useMergeRefs.call(void 0, ref, props.ref)
    });
    return _misc.removeUndefinedValues.call(void 0, props);
  }
);
var CollectionItem = _AWEKOM65cjs.forwardRef.call(void 0, function CollectionItem2(props) {
  const htmlProps = useCollectionItem(props);
  return _AWEKOM65cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useCollectionItem = useCollectionItem; exports.CollectionItem = CollectionItem;
