// @ariakit/react-core/cjs/__chunks/QA3YJPTA.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _PB5FNOHDcjs = require('./PB5FNOHD.cjs');



var _IHRK7HOTcjs = require('./IHRK7HOT.cjs');


var _GEDABTB5cjs = require('./GEDABTB5.cjs');




var _AWEKOM65cjs = require('./AWEKOM65.cjs');


var _OUWHJLQHcjs = require('./OUWHJLQH.cjs');



var _3IVNQNIVcjs = require('./3IVNQNIV.cjs');

// src/menubar/menubar.tsx
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useMenubar = _AWEKOM65cjs.createHook.call(void 0, 
  function useMenubar2(_a) {
    var _b = _a, {
      store: storeProp,
      composite = true,
      orientation: orientationProp,
      virtualFocus,
      focusLoop,
      rtl
    } = _b, props = _3IVNQNIVcjs.__objRest.call(void 0, _b, [
      "store",
      "composite",
      "orientation",
      "virtualFocus",
      "focusLoop",
      "rtl"
    ]);
    const context = _IHRK7HOTcjs.useMenubarProviderContext.call(void 0, );
    storeProp = storeProp || context;
    const store = _PB5FNOHDcjs.useMenubarStore.call(void 0, {
      store: storeProp,
      orientation: orientationProp,
      virtualFocus,
      focusLoop,
      rtl
    });
    const orientation = store.useState(
      (state) => !composite || state.orientation === "both" ? void 0 : state.orientation
    );
    props = _OUWHJLQHcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _IHRK7HOTcjs.MenubarScopedContextProvider, { value: store, children: element }),
      [store]
    );
    if (composite) {
      props = _3IVNQNIVcjs.__spreadValues.call(void 0, {
        role: "menubar",
        "aria-orientation": orientation
      }, props);
    }
    props = _GEDABTB5cjs.useComposite.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, { store, composite }, props));
    return props;
  }
);
var Menubar = _AWEKOM65cjs.forwardRef.call(void 0, function Menubar2(props) {
  const htmlProps = useMenubar(props);
  return _AWEKOM65cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useMenubar = useMenubar; exports.Menubar = Menubar;
