// @ariakit/react-core/cjs/__chunks/QC5YVEX2.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _BZ34A7Q7cjs = require('./BZ34A7Q7.cjs');



var _5XC2EBLVcjs = require('./5XC2EBLV.cjs');


var _AWEKOM65cjs = require('./AWEKOM65.cjs');

// src/combobox/combobox-context.tsx
var _react = require('react');
var ComboboxListRoleContext = _react.createContext.call(void 0, 
  void 0
);
var ctx = _AWEKOM65cjs.createStoreContext.call(void 0, 
  [_5XC2EBLVcjs.PopoverContextProvider, _BZ34A7Q7cjs.CompositeContextProvider],
  [_5XC2EBLVcjs.PopoverScopedContextProvider, _BZ34A7Q7cjs.CompositeScopedContextProvider]
);
var useComboboxContext = ctx.useContext;
var useComboboxScopedContext = ctx.useScopedContext;
var useComboboxProviderContext = ctx.useProviderContext;
var ComboboxContextProvider = ctx.ContextProvider;
var ComboboxScopedContextProvider = ctx.ScopedContextProvider;
var ComboboxItemValueContext = _react.createContext.call(void 0, 
  void 0
);
var ComboboxItemCheckedContext = _react.createContext.call(void 0, false);










exports.ComboboxListRoleContext = ComboboxListRoleContext; exports.useComboboxContext = useComboboxContext; exports.useComboboxScopedContext = useComboboxScopedContext; exports.useComboboxProviderContext = useComboboxProviderContext; exports.ComboboxContextProvider = ComboboxContextProvider; exports.ComboboxScopedContextProvider = ComboboxScopedContextProvider; exports.ComboboxItemValueContext = ComboboxItemValueContext; exports.ComboboxItemCheckedContext = ComboboxItemCheckedContext;
