// @ariakit/react-core/cjs/__chunks/SRE6U4QD.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";





var _QC5YVEX2cjs = require('./QC5YVEX2.cjs');


var _SY7NHEDFcjs = require('./SY7NHEDF.cjs');




var _AWEKOM65cjs = require('./AWEKOM65.cjs');






var _OUWHJLQHcjs = require('./OUWHJLQH.cjs');




var _3IVNQNIVcjs = require('./3IVNQNIV.cjs');

// src/combobox/combobox-list.tsx
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useComboboxList = _AWEKOM65cjs.createHook.call(void 0, 
  function useComboboxList2(_a) {
    var _b = _a, { store, alwaysVisible } = _b, props = _3IVNQNIVcjs.__objRest.call(void 0, _b, ["store", "alwaysVisible"]);
    const scopedContext = _QC5YVEX2cjs.useComboboxScopedContext.call(void 0, true);
    const context = _QC5YVEX2cjs.useComboboxContext.call(void 0, );
    store = store || context;
    const scopedContextSameStore = !!store && store === scopedContext;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "ComboboxList must receive a `store` prop or be wrapped in a ComboboxProvider component."
    );
    const ref = _react.useRef.call(void 0, null);
    const id = _OUWHJLQHcjs.useId.call(void 0, props.id);
    const mounted = store.useState("mounted");
    const hidden = _SY7NHEDFcjs.isHidden.call(void 0, mounted, props.hidden, alwaysVisible);
    const style = hidden ? _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, props.style), { display: "none" }) : props.style;
    const multiSelectable = store.useState(
      (state) => Array.isArray(state.selectedValue)
    );
    const role = _OUWHJLQHcjs.useAttribute.call(void 0, ref, "role", props.role);
    const isCompositeRole = role === "listbox" || role === "tree" || role === "grid";
    const ariaMultiSelectable = isCompositeRole ? multiSelectable || void 0 : void 0;
    const [hasListboxInside, setHasListboxInside] = _react.useState.call(void 0, false);
    const contentElement = store.useState("contentElement");
    _OUWHJLQHcjs.useSafeLayoutEffect.call(void 0, () => {
      if (!mounted) return;
      const element = ref.current;
      if (!element) return;
      if (contentElement !== element) return;
      const callback = () => {
        setHasListboxInside(!!element.querySelector("[role='listbox']"));
      };
      const observer = new MutationObserver(callback);
      observer.observe(element, {
        subtree: true,
        childList: true,
        attributeFilter: ["role"]
      });
      callback();
      return () => observer.disconnect();
    }, [mounted, contentElement]);
    if (!hasListboxInside) {
      props = _3IVNQNIVcjs.__spreadValues.call(void 0, {
        role: "listbox",
        "aria-multiselectable": ariaMultiSelectable
      }, props);
    }
    props = _OUWHJLQHcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _QC5YVEX2cjs.ComboboxScopedContextProvider, { value: store, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _QC5YVEX2cjs.ComboboxListRoleContext.Provider, { value: role, children: element }) }),
      [store, role]
    );
    const setContentElement = id && (!scopedContext || !scopedContextSameStore) ? store.setContentElement : null;
    props = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {
      id,
      hidden
    }, props), {
      ref: _OUWHJLQHcjs.useMergeRefs.call(void 0, setContentElement, ref, props.ref),
      style
    });
    return _misc.removeUndefinedValues.call(void 0, props);
  }
);
var ComboboxList = _AWEKOM65cjs.forwardRef.call(void 0, function ComboboxList2(props) {
  const htmlProps = useComboboxList(props);
  return _AWEKOM65cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useComboboxList = useComboboxList; exports.ComboboxList = ComboboxList;
