// @ariakit/react-core/cjs/__chunks/X7JQ6QHE.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _3LZQUHIBcjs = require('./3LZQUHIB.cjs');




var _AWEKOM65cjs = require('./AWEKOM65.cjs');



var _OUWHJLQHcjs = require('./OUWHJLQH.cjs');



var _3IVNQNIVcjs = require('./3IVNQNIV.cjs');

// src/heading/heading.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "h1";
var useHeading = _AWEKOM65cjs.createHook.call(void 0, 
  function useHeading2(props) {
    const ref = _react.useRef.call(void 0, null);
    const level = _react.useContext.call(void 0, _3LZQUHIBcjs.HeadingContext) || 1;
    const Element = `h${level}`;
    const tagName = _OUWHJLQHcjs.useTagName.call(void 0, ref, Element);
    const isNativeHeading = _react.useMemo.call(void 0, 
      () => !!tagName && /^h\d$/.test(tagName),
      [tagName]
    );
    props = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {
      render: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Element, {}),
      role: !isNativeHeading ? "heading" : void 0,
      "aria-level": !isNativeHeading ? level : void 0
    }, props), {
      ref: _OUWHJLQHcjs.useMergeRefs.call(void 0, ref, props.ref)
    });
    return props;
  }
);
var Heading = _AWEKOM65cjs.forwardRef.call(void 0, function Heading2(props) {
  const htmlProps = useHeading(props);
  return _AWEKOM65cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useHeading = useHeading; exports.Heading = Heading;
