// @wordpress/block-editor/build/components/block-tools/zoom-out-popover.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZoomOutPopover;
var _clsx = _interopRequireDefault(require("clsx"));
var _blockPopover = _interopRequireDefault(require("../block-popover"));
var _useBlockToolbarPopoverProps = _interopRequireDefault(require("./use-block-toolbar-popover-props"));
var _useSelectedBlockToolProps = _interopRequireDefault(require("./use-selected-block-tool-props"));
var _zoomOutToolbar = _interopRequireDefault(require("./zoom-out-toolbar"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * Internal dependencies
 */

function ZoomOutPopover({
  clientId,
  __unstableContentRef
}) {
  const {
    capturingClientId,
    isInsertionPointVisible,
    lastClientId,
    rootClientId
  } = (0, _useSelectedBlockToolProps.default)(clientId);
  const popoverProps = (0, _useBlockToolbarPopoverProps.default)({
    contentElement: __unstableContentRef?.current,
    clientId
  });

  // Override some of the popover props for the zoom-out toolbar.
  const props = {
    ...popoverProps,
    placement: 'left-start',
    flip: false,
    shift: true
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockPopover.default, {
    clientId: capturingClientId || clientId,
    bottomClientId: lastClientId,
    className: (0, _clsx.default)('zoom-out-toolbar-popover', {
      'is-insertion-point-visible': isInsertionPointVisible
    }),
    resize: false,
    ...props,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_zoomOutToolbar.default, {
      clientId: clientId,
      rootClientId: rootClientId
    })
  });
}
