// @wordpress/block-editor/build/components/block-tools/zoom-out-mode-inserters.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _data = require("@wordpress/data");
var _element = require("@wordpress/element");
var _inbetween = _interopRequireDefault(require("../block-popover/inbetween"));
var _zoomOutModeInserterButton = _interopRequireDefault(require("./zoom-out-mode-inserter-button"));
var _store = require("../../store");
var _lockUnlock = require("../../lock-unlock");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function ZoomOutModeInserters() {
  const [isReady, setIsReady] = (0, _element.useState)(false);
  const {
    hasSelection,
    blockInsertionPoint,
    blockOrder,
    blockInsertionPointVisible,
    setInserterIsOpened,
    sectionRootClientId,
    selectedBlockClientId,
    hoveredBlockClientId
  } = (0, _data.useSelect)(select => {
    const {
      getSettings,
      getBlockInsertionPoint,
      getBlockOrder,
      getSelectionStart,
      getSelectedBlockClientId,
      getHoveredBlockClientId,
      isBlockInsertionPointVisible
    } = select(_store.store);
    const {
      sectionRootClientId: root
    } = (0, _lockUnlock.unlock)(getSettings());
    return {
      hasSelection: !!getSelectionStart().clientId,
      blockInsertionPoint: getBlockInsertionPoint(),
      blockOrder: getBlockOrder(root),
      blockInsertionPointVisible: isBlockInsertionPointVisible(),
      sectionRootClientId: root,
      setInserterIsOpened: getSettings().__experimentalSetIsInserterOpened,
      selectedBlockClientId: getSelectedBlockClientId(),
      hoveredBlockClientId: getHoveredBlockClientId()
    };
  }, []);
  const {
    showInsertionPoint
  } = (0, _data.useDispatch)(_store.store);

  // Defer the initial rendering to avoid the jumps due to the animation.
  (0, _element.useEffect)(() => {
    const timeout = setTimeout(() => {
      setIsReady(true);
    }, 500);
    return () => {
      clearTimeout(timeout);
    };
  }, []);
  if (!isReady) {
    return null;
  }
  return [undefined, ...blockOrder].map((clientId, index) => {
    const shouldRenderInsertionPoint = blockInsertionPointVisible && blockInsertionPoint.index === index;
    const previousClientId = clientId;
    const nextClientId = blockOrder[index];
    const isSelected = hasSelection && (selectedBlockClientId === previousClientId || selectedBlockClientId === nextClientId);
    const isHovered = hoveredBlockClientId === previousClientId || hoveredBlockClientId === nextClientId;
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_inbetween.default, {
      previousClientId: previousClientId,
      nextClientId: nextClientId,
      children: [shouldRenderInsertionPoint && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        style: {
          borderRadius: '0',
          height: '12px',
          opacity: 1,
          transform: 'translateY(-50%)',
          width: '100%'
        },
        className: "block-editor-block-list__insertion-point-indicator"
      }), !shouldRenderInsertionPoint && /*#__PURE__*/(0, _jsxRuntime.jsx)(_zoomOutModeInserterButton.default, {
        isVisible: isSelected || isHovered,
        onClick: () => {
          setInserterIsOpened({
            rootClientId: sectionRootClientId,
            insertionIndex: index,
            tab: 'patterns',
            category: 'all'
          });
          showInsertionPoint(sectionRootClientId, index, {
            operation: 'insert'
          });
        }
      })]
    }, index);
  });
}
var _default = exports.default = ZoomOutModeInserters;
