// @wordpress/block-editor/build/components/inner-blocks/button-block-appender.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ButtonBlockAppender = void 0;
var _clsx = _interopRequireDefault(require("clsx"));
var _buttonBlockAppender = _interopRequireDefault(require("../button-block-appender"));
var _withClientId = _interopRequireDefault(require("./with-client-id"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * Internal dependencies
 */

const ButtonBlockAppender = ({
  clientId,
  showSeparator,
  isFloating,
  onAddBlock,
  isToggle
}) => {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_buttonBlockAppender.default, {
    className: (0, _clsx.default)({
      'block-list-appender__toggle': isToggle
    }),
    rootClientId: clientId,
    showSeparator: showSeparator,
    isFloating: isFloating,
    onAddBlock: onAddBlock
  });
};
exports.ButtonBlockAppender = ButtonBlockAppender;
var _default = exports.default = (0, _withClientId.default)(ButtonBlockAppender);
