// @ariakit/react-core/cjs/tag/tag.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _CINY3Q6Qcjs = require('../__chunks/CINY3Q6Q.cjs');
require('../__chunks/PYKELHKQ.cjs');
require('../__chunks/35VA3E7J.cjs');


var _MMWRTFBGcjs = require('../__chunks/MMWRTFBG.cjs');




var _LO2HIMYXcjs = require('../__chunks/LO2HIMYX.cjs');
require('../__chunks/BV5PE6PZ.cjs');
require('../__chunks/BZ34A7Q7.cjs');
require('../__chunks/55G3CDVQ.cjs');
require('../__chunks/EDR5S5CX.cjs');
require('../__chunks/DYIDES27.cjs');





var _AWEKOM65cjs = require('../__chunks/AWEKOM65.cjs');
require('../__chunks/PULY27HQ.cjs');





var _OUWHJLQHcjs = require('../__chunks/OUWHJLQH.cjs');
require('../__chunks/54XFMNDS.cjs');




var _3IVNQNIVcjs = require('../__chunks/3IVNQNIV.cjs');

// src/tag/tag.tsx
var _misc = require('@ariakit/core/utils/misc');
var _platform = require('@ariakit/core/utils/platform');
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useTag = _AWEKOM65cjs.createHook.call(void 0, function useTag2(_a) {
  var _b = _a, {
    store,
    value,
    getItem: getItemProp,
    removeOnKeyPress = true
  } = _b, props = _3IVNQNIVcjs.__objRest.call(void 0, _b, [
    "store",
    "value",
    "getItem",
    "removeOnKeyPress"
  ]);
  const context = _LO2HIMYXcjs.useTagContext.call(void 0, );
  store = store || context;
  _misc.invariant.call(void 0, 
    store,
    process.env.NODE_ENV !== "production" && "Tag must be wrapped in a TagList component."
  );
  const id = _OUWHJLQHcjs.useId.call(void 0, props.id);
  const touchDevice = _MMWRTFBGcjs.useTouchDevice.call(void 0, );
  const getItem = _react.useCallback.call(void 0, 
    (item) => {
      const nextItem = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, item), { value });
      if (getItemProp) {
        return getItemProp(nextItem);
      }
      return nextItem;
    },
    [value, getItemProp]
  );
  const onKeyDownProp = props.onKeyDown;
  const removeOnKeyPressProp = _OUWHJLQHcjs.useBooleanEvent.call(void 0, removeOnKeyPress);
  const onKeyDown = _OUWHJLQHcjs.useEvent.call(void 0, (event) => {
    onKeyDownProp == null ? void 0 : onKeyDownProp(event);
    if (event.defaultPrevented) return;
    const { inputElement } = store.getState();
    const isBackspace = event.key === "Backspace";
    const isRemoveKey = isBackspace || event.key === "Delete";
    if (isRemoveKey && removeOnKeyPressProp(event)) {
      event.preventDefault();
      store.removeValue(value);
      const nextId = isBackspace ? store.previous() || store.next() : store.next();
      store.move(nextId);
    }
    const isPrintableKey = !event.ctrlKey && !event.metaKey && event.key.length === 1;
    const pc = !_platform.isApple.call(void 0, );
    const v = event.key === "v" || event.key === "V";
    const mod = pc ? event.ctrlKey : event.metaKey;
    const isPaste = mod && v;
    if (isPrintableKey || isPaste) {
      inputElement == null ? void 0 : inputElement.focus();
    }
  });
  const [removeId, setRemoveId] = _react.useState.call(void 0, );
  props = _OUWHJLQHcjs.useWrapElement.call(void 0, 
    props,
    (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _LO2HIMYXcjs.TagValueContext.Provider, { value, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _LO2HIMYXcjs.TagRemoveIdContext.Provider, { value: setRemoveId, children: element }) }),
    [value]
  );
  props = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {
    id,
    role: !touchDevice ? "option" : "listitem",
    children: value,
    "aria-describedby": removeId
  }, props), {
    onKeyDown
  });
  props = _CINY3Q6Qcjs.useCompositeItem.call(void 0, _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {
    store
  }, props), {
    getItem
  }));
  return props;
});
var Tag = _AWEKOM65cjs.memo.call(void 0, 
  _AWEKOM65cjs.forwardRef.call(void 0, function Tag2(props) {
    const htmlProps = useTag(props);
    return _AWEKOM65cjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.Tag = Tag; exports.useTag = useTag;
