// @wordpress/dataviews/build/components/dataviews-bulk-actions-toolbar/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BulkActionsToolbar;
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
var _compose = require("@wordpress/compose");
var _data = require("@wordpress/data");
var _dataviewsBulkActions = require("../dataviews-bulk-actions");
var _dataviewsContext = _interopRequireDefault(require("../dataviews-context"));
var _dataviewsItemActions = require("../dataviews-item-actions");
var _constants = require("../../constants");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const SNACKBAR_VARIANTS = {
  init: {
    bottom: -48
  },
  open: {
    bottom: 24,
    transition: {
      bottom: {
        type: 'tween',
        duration: 0.2,
        ease: [0, 0, 0.2, 1]
      }
    }
  },
  exit: {
    opacity: 0,
    bottom: 24,
    transition: {
      opacity: {
        type: 'tween',
        duration: 0.2,
        ease: [0, 0, 0.2, 1]
      }
    }
  }
};
function ActionTrigger({
  action,
  onClick,
  isBusy,
  items
}) {
  const label = typeof action.label === 'string' ? action.label : action.label(items);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.ToolbarButton, {
    disabled: isBusy,
    label: label,
    icon: action.icon,
    isDestructive: action.isDestructive,
    size: "compact",
    onClick: onClick,
    isBusy: isBusy,
    tooltipPosition: "top"
  });
}
const EMPTY_ARRAY = [];
function ActionButton({
  action,
  selectedItems,
  actionInProgress,
  setActionInProgress
}) {
  const registry = (0, _data.useRegistry)();
  const selectedEligibleItems = (0, _element.useMemo)(() => {
    return selectedItems.filter(item => {
      return !action.isEligible || action.isEligible(item);
    });
  }, [action, selectedItems]);
  if ('RenderModal' in action) {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_dataviewsItemActions.ActionWithModal, {
      action: action,
      items: selectedEligibleItems,
      ActionTrigger: ActionTrigger
    }, action.id);
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(ActionTrigger, {
    action: action,
    onClick: () => {
      setActionInProgress(action.id);
      action.callback(selectedItems, {
        registry
      });
    },
    items: selectedEligibleItems,
    isBusy: actionInProgress === action.id
  }, action.id);
}
function renderToolbarContent(selection, actionsToShow, selectedItems, actionInProgress, setActionInProgress, onChangeSelection) {
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.ToolbarGroup, {
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        className: "dataviews-bulk-actions-toolbar__selection-count",
        children: selection.length === 1 ? (0, _i18n.__)('1 item selected') : (0, _i18n.sprintf)(
        // translators: %s: Total number of selected items.
        (0, _i18n._n)('%s item selected', '%s items selected', selection.length), selection.length)
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.ToolbarGroup, {
      children: actionsToShow.map(action => {
        return /*#__PURE__*/(0, _jsxRuntime.jsx)(ActionButton, {
          action: action,
          selectedItems: selectedItems,
          actionInProgress: actionInProgress,
          setActionInProgress: setActionInProgress
        }, action.id);
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.ToolbarGroup, {
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.ToolbarButton, {
        icon: _icons.closeSmall,
        showTooltip: true,
        tooltipPosition: "top",
        label: (0, _i18n.__)('Cancel'),
        disabled: !!actionInProgress,
        onClick: () => {
          onChangeSelection(EMPTY_ARRAY);
        }
      })
    })]
  });
}
function ToolbarContent({
  selection,
  actionsToShow,
  selectedItems,
  onChangeSelection
}) {
  const [actionInProgress, setActionInProgress] = (0, _element.useState)(null);
  const buttons = (0, _element.useRef)(null);
  if (!actionInProgress) {
    if (buttons.current) {
      buttons.current = null;
    }
    return renderToolbarContent(selection, actionsToShow, selectedItems, actionInProgress, setActionInProgress, onChangeSelection);
  } else if (!buttons.current) {
    buttons.current = renderToolbarContent(selection, actionsToShow, selectedItems, actionInProgress, setActionInProgress, onChangeSelection);
  }
  return buttons.current;
}
function _BulkActionsToolbar() {
  const {
    data,
    selection,
    actions = EMPTY_ARRAY,
    onChangeSelection,
    getItemId
  } = (0, _element.useContext)(_dataviewsContext.default);
  const isReducedMotion = (0, _compose.useReducedMotion)();
  const selectedItems = (0, _element.useMemo)(() => {
    return data.filter(item => selection.includes(getItemId(item)));
  }, [selection, data, getItemId]);
  const actionsToShow = (0, _element.useMemo)(() => actions.filter(action => {
    return action.supportsBulk && action.icon && selectedItems.some(item => !action.isEligible || action.isEligible(item));
  }), [actions, selectedItems]);
  if (selection && selection.length === 0 || actionsToShow.length === 0) {
    return null;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__unstableAnimatePresence, {
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__unstableMotion.div, {
      layout: !isReducedMotion // See https://www.framer.com/docs/animation/#layout-animations
      ,
      initial: "init",
      animate: "open",
      exit: "exit",
      variants: isReducedMotion ? undefined : SNACKBAR_VARIANTS,
      className: "dataviews-bulk-actions-toolbar",
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Toolbar, {
        label: (0, _i18n.__)('Bulk actions'),
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: "dataviews-bulk-actions-toolbar__wrapper",
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(ToolbarContent, {
            selection: selection,
            actionsToShow: actionsToShow,
            selectedItems: selectedItems,
            onChangeSelection: onChangeSelection
          })
        })
      })
    })
  });
}
function BulkActionsToolbar() {
  const {
    data,
    actions = [],
    view
  } = (0, _element.useContext)(_dataviewsContext.default);
  const hasPossibleBulkAction = (0, _dataviewsBulkActions.useSomeItemHasAPossibleBulkAction)(actions, data);
  if (![_constants.LAYOUT_TABLE, _constants.LAYOUT_GRID].includes(view.type) || !hasPossibleBulkAction) {
    return null;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_BulkActionsToolbar, {});
}
