// @wordpress/dataviews/build-module/field-types/integer.js
/**
 * WordPress dependencies
 */
import { __experimentalNumberControl as NumberControl, SelectControl } from '@wordpress/components';
import { useCallback } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { jsx as _jsx } from "react/jsx-runtime";
function sort(a, b, direction) {
  return direction === 'asc' ? a - b : b - a;
}
function isValid(value, context) {
  // TODO: this implicitely means the value is required.
  if (value === '') {
    return false;
  }
  if (!Number.isInteger(Number(value))) {
    return false;
  }
  if (context?.elements) {
    const validValues = context?.elements.map(f => f.value);
    if (!validValues.includes(Number(value))) {
      return false;
    }
  }
  return true;
}
function Edit({
  data,
  field,
  onChange,
  hideLabelFromVision
}) {
  var _field$getValue;
  const {
    id,
    label,
    description
  } = field;
  const value = (_field$getValue = field.getValue({
    item: data
  })) !== null && _field$getValue !== void 0 ? _field$getValue : '';
  const onChangeControl = useCallback(newValue => onChange(prevItem => ({
    ...prevItem,
    [id]: newValue
  })), [id, onChange]);
  if (field.elements) {
    const elements = [
    /*
     * Value can be undefined when:
     *
     * - the field is not required
     * - in bulk editing
     *
     */
    {
      label: __('Select item'),
      value: ''
    }, ...field.elements];
    return /*#__PURE__*/_jsx(SelectControl, {
      label: label,
      value: value,
      options: elements,
      onChange: onChangeControl,
      __next40pxDefaultSize: true,
      __nextHasNoMarginBottom: true,
      hideLabelFromVision: hideLabelFromVision
    });
  }
  return /*#__PURE__*/_jsx(NumberControl, {
    label: label,
    help: description,
    value: value,
    onChange: onChangeControl,
    __next40pxDefaultSize: true,
    hideLabelFromVision: hideLabelFromVision
  });
}
export default {
  sort,
  isValid,
  Edit
};
