// @wordpress/dataviews/build-module/field-types/text.js
/**
 * WordPress dependencies
 */
import { SelectControl, TextControl } from '@wordpress/components';
import { useCallback } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { jsx as _jsx } from "react/jsx-runtime";
function sort(valueA, valueB, direction) {
  return direction === 'asc' ? valueA.localeCompare(valueB) : valueB.localeCompare(valueA);
}
function isValid(value, context) {
  if (context?.elements) {
    const validValues = context?.elements?.map(f => f.value);
    if (!validValues.includes(value)) {
      return false;
    }
  }
  return true;
}
function Edit({
  data,
  field,
  onChange,
  hideLabelFromVision
}) {
  const {
    id,
    label,
    placeholder
  } = field;
  const value = field.getValue({
    item: data
  });
  const onChangeControl = useCallback(newValue => onChange(prevItem => ({
    ...prevItem,
    [id]: newValue
  })), [id, onChange]);
  if (field.elements) {
    const elements = [
    /*
     * Value can be undefined when:
     *
     * - the field is not required
     * - in bulk editing
     *
     */
    {
      label: __('Select item'),
      value: ''
    }, ...field.elements];
    return /*#__PURE__*/_jsx(SelectControl, {
      label: label,
      value: value,
      options: elements,
      onChange: onChangeControl,
      __next40pxDefaultSize: true,
      __nextHasNoMarginBottom: true,
      hideLabelFromVision: hideLabelFromVision
    });
  }
  return /*#__PURE__*/_jsx(TextControl, {
    label: label,
    placeholder: placeholder,
    value: value !== null && value !== void 0 ? value : '',
    onChange: onChangeControl,
    __next40pxDefaultSize: true,
    __nextHasNoMarginBottom: true,
    hideLabelFromVision: hideLabelFromVision
  });
}
export default {
  sort,
  isValid,
  Edit
};
