// @wordpress/editor/build/components/block-manager/checklist.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _blockEditor = require("@wordpress/block-editor");
var _components = require("@wordpress/components");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

function BlockTypesChecklist({
  blockTypes,
  value,
  onItemChange
}) {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("ul", {
    className: "editor-block-manager__checklist",
    children: blockTypes.map(blockType => /*#__PURE__*/(0, _jsxRuntime.jsxs)("li", {
      className: "editor-block-manager__checklist-item",
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.CheckboxControl, {
        __nextHasNoMarginBottom: true,
        label: blockType.title,
        checked: value.includes(blockType.name),
        onChange: (...args) => onItemChange(blockType.name, ...args)
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.BlockIcon, {
        icon: blockType.icon
      })]
    }, blockType.name))
  });
}
var _default = exports.default = BlockTypesChecklist;
