// @ariakit/react-core/cjs/select/select-item.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _XV4AOZNFcjs = require('../__chunks/XV4AOZNF.cjs');



var _I52OTQQ2cjs = require('../__chunks/I52OTQQ2.cjs');


var _CINY3Q6Qcjs = require('../__chunks/CINY3Q6Q.cjs');
require('../__chunks/PYKELHKQ.cjs');
require('../__chunks/35VA3E7J.cjs');
require('../__chunks/BV5PE6PZ.cjs');
require('../__chunks/BZ34A7Q7.cjs');
require('../__chunks/55G3CDVQ.cjs');
require('../__chunks/5XC2EBLV.cjs');
require('../__chunks/CQ3ILEKO.cjs');
require('../__chunks/NRLE3XE7.cjs');
require('../__chunks/EDR5S5CX.cjs');
require('../__chunks/DYIDES27.cjs');





var _AWEKOM65cjs = require('../__chunks/AWEKOM65.cjs');
require('../__chunks/PULY27HQ.cjs');





var _OUWHJLQHcjs = require('../__chunks/OUWHJLQH.cjs');
require('../__chunks/54XFMNDS.cjs');




var _3IVNQNIVcjs = require('../__chunks/3IVNQNIV.cjs');

// src/select/select-item.tsx
var _dom = require('@ariakit/core/utils/dom');
var _events = require('@ariakit/core/utils/events');
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
function isSelected(storeValue, itemValue) {
  if (itemValue == null) return;
  if (storeValue == null) return false;
  if (Array.isArray(storeValue)) {
    return storeValue.includes(itemValue);
  }
  return storeValue === itemValue;
}
var useSelectItem = _AWEKOM65cjs.createHook.call(void 0, 
  function useSelectItem2(_a) {
    var _b = _a, {
      store,
      value,
      getItem: getItemProp,
      hideOnClick,
      setValueOnClick = value != null,
      preventScrollOnKeyDown = true,
      focusOnHover = true
    } = _b, props = _3IVNQNIVcjs.__objRest.call(void 0, _b, [
      "store",
      "value",
      "getItem",
      "hideOnClick",
      "setValueOnClick",
      "preventScrollOnKeyDown",
      "focusOnHover"
    ]);
    var _a2;
    const context = _I52OTQQ2cjs.useSelectScopedContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "SelectItem must be wrapped in a SelectList or SelectPopover component."
    );
    const id = _OUWHJLQHcjs.useId.call(void 0, props.id);
    const disabled = _misc.disabledFromProps.call(void 0, props);
    const getItem = _react.useCallback.call(void 0, 
      (item) => {
        const nextItem = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, item), {
          value: disabled ? void 0 : value,
          children: value
        });
        if (getItemProp) {
          return getItemProp(nextItem);
        }
        return nextItem;
      },
      [disabled, value, getItemProp]
    );
    const multiSelectable = store.useState(
      (state) => Array.isArray(state.value)
    );
    hideOnClick = hideOnClick != null ? hideOnClick : value != null && !multiSelectable;
    const onClickProp = props.onClick;
    const setValueOnClickProp = _OUWHJLQHcjs.useBooleanEvent.call(void 0, setValueOnClick);
    const hideOnClickProp = _OUWHJLQHcjs.useBooleanEvent.call(void 0, hideOnClick);
    const onClick = _OUWHJLQHcjs.useEvent.call(void 0, (event) => {
      onClickProp == null ? void 0 : onClickProp(event);
      if (event.defaultPrevented) return;
      if (_events.isDownloading.call(void 0, event)) return;
      if (_events.isOpeningInNewTab.call(void 0, event)) return;
      if (setValueOnClickProp(event) && value != null) {
        store == null ? void 0 : store.setValue((prevValue) => {
          if (!Array.isArray(prevValue)) return value;
          if (prevValue.includes(value)) {
            return prevValue.filter((v) => v !== value);
          }
          return [...prevValue, value];
        });
      }
      if (hideOnClickProp(event)) {
        store == null ? void 0 : store.hide();
      }
    });
    const selected = store.useState((state) => isSelected(state.value, value));
    props = _OUWHJLQHcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _I52OTQQ2cjs.SelectItemCheckedContext.Provider, { value: selected != null ? selected : false, children: element }),
      [selected]
    );
    const listElement = store.useState("listElement");
    const autoFocus = store.useState((state) => {
      if (value == null) return false;
      if (state.value == null) return false;
      if (state.activeId !== id && (store == null ? void 0 : store.item(state.activeId))) return false;
      if (Array.isArray(state.value)) {
        return state.value[state.value.length - 1] === value;
      }
      return state.value === value;
    });
    props = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {
      id,
      role: _dom.getPopupItemRole.call(void 0, listElement),
      "aria-selected": selected,
      children: value
    }, props), {
      autoFocus: (_a2 = props.autoFocus) != null ? _a2 : autoFocus,
      onClick
    });
    props = _CINY3Q6Qcjs.useCompositeItem.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {
      store,
      getItem,
      preventScrollOnKeyDown
    }, props));
    const focusOnHoverProp = _OUWHJLQHcjs.useBooleanEvent.call(void 0, focusOnHover);
    props = _XV4AOZNFcjs.useCompositeHover.call(void 0, _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {
      store
    }, props), {
      // We have to disable focusOnHover when the popup is closed, otherwise
      // the active item will change to null (the container) when the popup is
      // closed by clicking on an item.
      focusOnHover(event) {
        if (!focusOnHoverProp(event)) return false;
        const state = store == null ? void 0 : store.getState();
        return !!(state == null ? void 0 : state.open);
      }
    }));
    return props;
  }
);
var SelectItem = _AWEKOM65cjs.memo.call(void 0, 
  _AWEKOM65cjs.forwardRef.call(void 0, function SelectItem2(props) {
    const htmlProps = useSelectItem(props);
    return _AWEKOM65cjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.SelectItem = SelectItem; exports.useSelectItem = useSelectItem;
