// @wordpress/dataviews/build-module/components/dataviews-view-config/index.js
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
import { Button, Popover, __experimentalToggleGroupControl as ToggleGroupControl, __experimentalToggleGroupControlOption as ToggleGroupControlOption, __experimentalToggleGroupControlOptionIcon as ToggleGroupControlOptionIcon, SelectControl, __experimentalItemGroup as ItemGroup, __experimentalItem as Item, __experimentalGrid as Grid, __experimentalVStack as VStack, __experimentalHStack as HStack, __experimentalHeading as Heading, __experimentalText as Text, privateApis as componentsPrivateApis } from '@wordpress/components';
import { __, _x } from '@wordpress/i18n';
import { memo, useContext, useState, useMemo } from '@wordpress/element';
import { cog, seen, unseen } from '@wordpress/icons';
import warning from '@wordpress/warning';

/**
 * Internal dependencies
 */
import { SORTING_DIRECTIONS, sortIcons, sortLabels } from '../../constants';
import { VIEW_LAYOUTS, getMandatoryFields } from '../../dataviews-layouts';
import DataViewsContext from '../dataviews-context';
import { unlock } from '../../lock-unlock';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
import { Fragment as _Fragment } from "react/jsx-runtime";
const {
  DropdownMenuV2: DropdownMenu,
  DropdownMenuRadioItemV2: DropdownMenuRadioItem,
  DropdownMenuItemLabelV2: DropdownMenuItemLabel
} = unlock(componentsPrivateApis);
function ViewTypeMenu({
  defaultLayouts = {
    list: {},
    grid: {},
    table: {}
  }
}) {
  const {
    view,
    onChangeView
  } = useContext(DataViewsContext);
  const availableLayouts = Object.keys(defaultLayouts);
  if (availableLayouts.length <= 1) {
    return null;
  }
  const activeView = VIEW_LAYOUTS.find(v => view.type === v.type);
  return /*#__PURE__*/_jsx(DropdownMenu, {
    trigger: /*#__PURE__*/_jsx(Button, {
      size: "compact",
      icon: activeView?.icon,
      label: __('Layout')
    }),
    children: availableLayouts.map(layout => {
      const config = VIEW_LAYOUTS.find(v => v.type === layout);
      if (!config) {
        return null;
      }
      return /*#__PURE__*/_jsx(DropdownMenuRadioItem, {
        value: layout,
        name: "view-actions-available-view",
        checked: layout === view.type,
        hideOnClick: true,
        onChange: e => {
          switch (e.target.value) {
            case 'list':
            case 'grid':
            case 'table':
              return onChangeView({
                ...view,
                type: e.target.value,
                ...defaultLayouts[e.target.value]
              });
          }
          globalThis.SCRIPT_DEBUG === true ? warning('Invalid dataview') : void 0;
        },
        children: /*#__PURE__*/_jsx(DropdownMenuItemLabel, {
          children: config.label
        })
      }, layout);
    })
  });
}
function SortFieldControl() {
  const {
    view,
    fields,
    onChangeView
  } = useContext(DataViewsContext);
  const orderOptions = useMemo(() => {
    const sortableFields = fields.filter(field => field.enableSorting !== false);
    return sortableFields.map(field => {
      return {
        label: field.label,
        value: field.id
      };
    });
  }, [fields]);
  return /*#__PURE__*/_jsx(SelectControl, {
    __nextHasNoMarginBottom: true,
    __next40pxDefaultSize: true,
    label: __('Sort by'),
    value: view.sort?.field,
    options: orderOptions,
    onChange: value => {
      onChangeView({
        ...view,
        sort: {
          direction: view?.sort?.direction || 'desc',
          field: value
        }
      });
    }
  });
}
function SortDirectionControl() {
  const {
    view,
    onChangeView
  } = useContext(DataViewsContext);
  return /*#__PURE__*/_jsx(ToggleGroupControl, {
    className: "dataviews-view-config__sort-direction",
    __nextHasNoMarginBottom: true,
    __next40pxDefaultSize: true,
    isBlock: true,
    label: __('Order'),
    value: view.sort?.direction || 'desc',
    disabled: !view?.sort?.field,
    onChange: newDirection => {
      if (!view?.sort?.field) {
        return;
      }
      if (newDirection === 'asc' || newDirection === 'desc') {
        onChangeView({
          ...view,
          sort: {
            direction: newDirection,
            field: view.sort.field
          }
        });
        return;
      }
      globalThis.SCRIPT_DEBUG === true ? warning('Invalid direction') : void 0;
    },
    children: SORTING_DIRECTIONS.map(direction => {
      return /*#__PURE__*/_jsx(ToggleGroupControlOptionIcon, {
        value: direction,
        icon: sortIcons[direction],
        label: sortLabels[direction]
      }, direction);
    })
  });
}
const PAGE_SIZE_VALUES = [10, 20, 50, 100];
function ItemsPerPageControl() {
  const {
    view,
    onChangeView
  } = useContext(DataViewsContext);
  return /*#__PURE__*/_jsx(ToggleGroupControl, {
    __nextHasNoMarginBottom: true,
    __next40pxDefaultSize: true,
    isBlock: true,
    label: __('Items per page'),
    value: view.perPage || 10,
    disabled: !view?.sort?.field,
    onChange: newItemsPerPage => {
      const newItemsPerPageNumber = typeof newItemsPerPage === 'number' || newItemsPerPage === undefined ? newItemsPerPage : parseInt(newItemsPerPage, 10);
      onChangeView({
        ...view,
        perPage: newItemsPerPageNumber,
        page: 1
      });
    },
    children: PAGE_SIZE_VALUES.map(value => {
      return /*#__PURE__*/_jsx(ToggleGroupControlOption, {
        value: value,
        label: value.toString()
      }, value);
    })
  });
}
function FieldControl() {
  const {
    view,
    fields,
    onChangeView
  } = useContext(DataViewsContext);
  const mandatoryFields = getMandatoryFields(view);
  const hidableFields = fields.filter(field => field.enableHiding !== false && !mandatoryFields.includes(field.id));
  const viewFields = view.fields || fields.map(field => field.id);
  if (!hidableFields?.length) {
    return null;
  }
  return /*#__PURE__*/_jsx(ItemGroup, {
    isBordered: true,
    isSeparated: true,
    children: hidableFields?.map(field => {
      const isVisible = viewFields.includes(field.id);
      return /*#__PURE__*/_jsx(Item, {
        children: /*#__PURE__*/_jsxs(HStack, {
          expanded: true,
          children: [/*#__PURE__*/_jsx("span", {
            children: field.label
          }), /*#__PURE__*/_jsx(Button, {
            className: "'dataviews-view-config__field-control-button",
            size: "compact",
            onClick: () => onChangeView({
              ...view,
              fields: isVisible ? viewFields.filter(id => id !== field.id) : [...viewFields, field.id]
            }),
            icon: isVisible ? seen : unseen,
            label: isVisible ? __('Hide field') : __('Show field')
          })]
        })
      }, field.id);
    })
  });
}
function SettingsSection({
  title,
  description,
  children
}) {
  return /*#__PURE__*/_jsxs(Grid, {
    columns: 12,
    className: "dataviews-settings-section",
    gap: 4,
    children: [/*#__PURE__*/_jsxs("div", {
      className: "dataviews-settings-section__sidebar",
      children: [/*#__PURE__*/_jsx(Heading, {
        level: 2,
        className: "dataviews-settings-section__title",
        children: title
      }), description && /*#__PURE__*/_jsx(Text, {
        variant: "muted",
        className: "dataviews-settings-section__description",
        children: description
      })]
    }), /*#__PURE__*/_jsx(Grid, {
      columns: 8,
      gap: 4,
      className: "dataviews-settings-section__content",
      children: children
    })]
  });
}
function DataviewsViewConfigContent() {
  return /*#__PURE__*/_jsxs(VStack, {
    className: "dataviews-view-config",
    spacing: 6,
    children: [/*#__PURE__*/_jsxs(SettingsSection, {
      title: __('Appearance'),
      children: [/*#__PURE__*/_jsxs(HStack, {
        expanded: true,
        className: "is-divided-in-two",
        children: [/*#__PURE__*/_jsx(SortFieldControl, {}), /*#__PURE__*/_jsx(SortDirectionControl, {})]
      }), /*#__PURE__*/_jsx(ItemsPerPageControl, {})]
    }), /*#__PURE__*/_jsx(SettingsSection, {
      title: __('Properties'),
      children: /*#__PURE__*/_jsx(FieldControl, {})
    })]
  });
}
function _DataViewsViewConfig({
  defaultLayouts = {
    list: {},
    grid: {},
    table: {}
  }
}) {
  const [isShowingViewPopover, setIsShowingViewPopover] = useState(false);
  return /*#__PURE__*/_jsxs(_Fragment, {
    children: [/*#__PURE__*/_jsx(ViewTypeMenu, {
      defaultLayouts: defaultLayouts
    }), /*#__PURE__*/_jsxs("div", {
      children: [/*#__PURE__*/_jsx(Button, {
        size: "compact",
        icon: cog,
        label: _x('View options', 'View is used as a noun'),
        onClick: () => setIsShowingViewPopover(true)
      }), isShowingViewPopover && /*#__PURE__*/_jsx(Popover, {
        placement: "bottom-end",
        onClose: () => {
          setIsShowingViewPopover(false);
        },
        focusOnMount: true,
        children: /*#__PURE__*/_jsx(DataviewsViewConfigContent, {})
      })]
    })]
  });
}
const DataViewsViewConfig = memo(_DataViewsViewConfig);
export default DataViewsViewConfig;
