// @wordpress/dataviews/build/dataviews-layouts/list/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ViewList;
var _clsx = _interopRequireDefault(require("clsx"));
var _compose = require("@wordpress/compose");
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
var _data = require("@wordpress/data");
var _lockUnlock = require("../../lock-unlock");
var _dataviewsItemActions = require("../../components/dataviews-item-actions");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

// Import CompositeStore type, which is not exported from @wordpress/components.
// eslint-disable-next-line no-restricted-imports

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  useCompositeStoreV2: useCompositeStore,
  CompositeV2: Composite,
  CompositeItemV2: CompositeItem,
  CompositeRowV2: CompositeRow,
  DropdownMenuV2: DropdownMenu
} = (0, _lockUnlock.unlock)(_components.privateApis);
function ListItem({
  actions,
  id,
  isSelected,
  item,
  mediaField,
  onSelect,
  primaryField,
  store,
  visibleFields
}) {
  const registry = (0, _data.useRegistry)();
  const itemRef = (0, _element.useRef)(null);
  const labelId = `${id}-label`;
  const descriptionId = `${id}-description`;
  const [isHovered, setIsHovered] = (0, _element.useState)(false);
  const handleMouseEnter = () => {
    setIsHovered(true);
  };
  const handleMouseLeave = () => {
    setIsHovered(false);
  };
  (0, _element.useEffect)(() => {
    if (isSelected) {
      itemRef.current?.scrollIntoView({
        behavior: 'auto',
        block: 'nearest',
        inline: 'nearest'
      });
    }
  }, [isSelected]);
  const {
    primaryAction,
    eligibleActions
  } = (0, _element.useMemo)(() => {
    // If an action is eligible for all items, doesn't need
    // to provide the `isEligible` function.
    const _eligibleActions = actions.filter(action => !action.isEligible || action.isEligible(item));
    const _primaryActions = _eligibleActions.filter(action => action.isPrimary && !!action.icon);
    return {
      primaryAction: _primaryActions?.[0],
      eligibleActions: _eligibleActions
    };
  }, [actions, item]);
  const [isModalOpen, setIsModalOpen] = (0, _element.useState)(false);
  const primaryActionLabel = primaryAction && (typeof primaryAction.label === 'string' ? primaryAction.label : primaryAction.label([item]));
  const renderedMediaField = mediaField?.render ? /*#__PURE__*/(0, _jsxRuntime.jsx)(mediaField.render, {
    item: item
  }) : /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: "dataviews-view-list__media-placeholder"
  });
  const renderedPrimaryField = primaryField?.render ? /*#__PURE__*/(0, _jsxRuntime.jsx)(primaryField.render, {
    item: item
  }) : null;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(CompositeRow, {
    ref: itemRef,
    render: /*#__PURE__*/(0, _jsxRuntime.jsx)("li", {}),
    role: "row",
    className: (0, _clsx.default)({
      'is-selected': isSelected,
      'is-hovered': isHovered
    }),
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
      className: "dataviews-view-list__item-wrapper",
      alignment: "center",
      spacing: 0,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        role: "gridcell",
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(CompositeItem, {
          store: store,
          render: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {}),
          role: "button",
          id: id,
          "aria-pressed": isSelected,
          "aria-labelledby": labelId,
          "aria-describedby": descriptionId,
          className: "dataviews-view-list__item",
          onClick: () => onSelect(item),
          children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
            spacing: 3,
            justify: "start",
            alignment: "flex-start",
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
              className: "dataviews-view-list__media-wrapper",
              children: renderedMediaField
            }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalVStack, {
              spacing: 1,
              className: "dataviews-view-list__field-wrapper",
              children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                className: "dataviews-view-list__primary-field",
                id: labelId,
                children: renderedPrimaryField
              }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
                className: "dataviews-view-list__fields",
                id: descriptionId,
                children: visibleFields.map(field => /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
                  className: "dataviews-view-list__field",
                  children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.VisuallyHidden, {
                    as: "span",
                    className: "dataviews-view-list__field-label",
                    children: field.label
                  }), /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                    className: "dataviews-view-list__field-value",
                    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(field.render, {
                      item: item
                    })
                  })]
                }, field.id))
              })]
            })]
          })
        })
      }), eligibleActions?.length > 0 && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
        spacing: 3,
        justify: "flex-end",
        className: "dataviews-view-list__item-actions",
        style: {
          flexShrink: '0',
          width: 'auto'
        },
        children: [primaryAction && 'RenderModal' in primaryAction && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          role: "gridcell",
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(CompositeItem, {
            store: store,
            render: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
              label: primaryActionLabel,
              icon: primaryAction.icon,
              isDestructive: primaryAction.isDestructive,
              size: "small",
              onClick: () => setIsModalOpen(true)
            }),
            children: isModalOpen && /*#__PURE__*/(0, _jsxRuntime.jsx)(_dataviewsItemActions.ActionModal, {
              action: primaryAction,
              items: [item],
              closeModal: () => setIsModalOpen(false)
            })
          })
        }), primaryAction && !('RenderModal' in primaryAction) && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          role: "gridcell",
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(CompositeItem, {
            store: store,
            render: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
              label: primaryActionLabel,
              icon: primaryAction.icon,
              isDestructive: primaryAction.isDestructive,
              size: "small",
              onClick: () => {
                primaryAction.callback([item], {
                  registry
                });
              }
            })
          })
        }, primaryAction.id), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          role: "gridcell",
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenu, {
            trigger: /*#__PURE__*/(0, _jsxRuntime.jsx)(CompositeItem, {
              store: store,
              render: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
                size: "small",
                icon: _icons.moreVertical,
                label: (0, _i18n.__)('Actions'),
                accessibleWhenDisabled: true,
                disabled: !actions.length,
                onKeyDown: event => {
                  if (event.key === 'ArrowDown') {
                    // Prevent the default behaviour (open dropdown menu) and go down.
                    event.preventDefault();
                    store.move(store.down());
                  }
                  if (event.key === 'ArrowUp') {
                    // Prevent the default behavior (open dropdown menu) and go up.
                    event.preventDefault();
                    store.move(store.up());
                  }
                }
              })
            }),
            placement: "bottom-end",
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_dataviewsItemActions.ActionsDropdownMenuGroup, {
              actions: eligibleActions,
              item: item
            })
          })
        })]
      })]
    })
  });
}
function ViewList(props) {
  const {
    actions,
    data,
    fields,
    getItemId,
    isLoading,
    onChangeSelection,
    selection,
    view
  } = props;
  const baseId = (0, _compose.useInstanceId)(ViewList, 'view-list');
  const selectedItem = data?.findLast(item => selection.includes(getItemId(item)));
  const mediaField = fields.find(field => field.id === view.layout?.mediaField);
  const primaryField = fields.find(field => field.id === view.layout?.primaryField);
  const viewFields = view.fields || fields.map(field => field.id);
  const visibleFields = fields.filter(field => viewFields.includes(field.id) && ![view.layout?.primaryField, view.layout?.mediaField].includes(field.id));
  const onSelect = item => onChangeSelection([getItemId(item)]);
  const getItemDomId = (0, _element.useCallback)(item => item ? `${baseId}-${getItemId(item)}` : undefined, [baseId, getItemId]);
  const store = useCompositeStore({
    defaultActiveId: getItemDomId(selectedItem)
  });

  // Manage focused item, when the active one is removed from the list.
  const isActiveIdInList = store.useState(state => state.items.some(item => item.id === state.activeId));
  (0, _element.useEffect)(() => {
    if (!isActiveIdInList) {
      // Prefer going down, except if there is no item below (last item), then go up (last item in list).
      if (store.down()) {
        store.move(store.down());
      } else if (store.up()) {
        store.move(store.up());
      }
    }
  }, [isActiveIdInList]);
  const hasData = data?.length;
  if (!hasData) {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: (0, _clsx.default)({
        'dataviews-loading': isLoading,
        'dataviews-no-results': !hasData && !isLoading
      }),
      children: !hasData && /*#__PURE__*/(0, _jsxRuntime.jsx)("p", {
        children: isLoading ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Spinner, {}) : (0, _i18n.__)('No results')
      })
    });
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(Composite, {
    id: baseId,
    render: /*#__PURE__*/(0, _jsxRuntime.jsx)("ul", {}),
    className: "dataviews-view-list",
    role: "grid",
    store: store,
    children: data.map(item => {
      const id = getItemDomId(item);
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(ListItem, {
        id: id,
        actions: actions,
        item: item,
        isSelected: item === selectedItem,
        onSelect: onSelect,
        mediaField: mediaField,
        primaryField: primaryField,
        store: store,
        visibleFields: visibleFields
      }, id);
    })
  });
}
