// @wordpress/block-editor/build/components/provider/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ExperimentalBlockEditorProvider = exports.BlockEditorProvider = void 0;
var _data = require("@wordpress/data");
var _element = require("@wordpress/element");
var _components = require("@wordpress/components");
var _withRegistryProvider = _interopRequireDefault(require("./with-registry-provider"));
var _useBlockSync = _interopRequireDefault(require("./use-block-sync"));
var _store = require("../../store");
var _blockRefsProvider = require("./block-refs-provider");
var _lockUnlock = require("../../lock-unlock");
var _keyboardShortcuts = _interopRequireDefault(require("../keyboard-shortcuts"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/** @typedef {import('@wordpress/data').WPDataRegistry} WPDataRegistry */

const ExperimentalBlockEditorProvider = exports.ExperimentalBlockEditorProvider = (0, _withRegistryProvider.default)(props => {
  const {
    children,
    settings,
    stripExperimentalSettings = false
  } = props;
  const {
    __experimentalUpdateSettings
  } = (0, _lockUnlock.unlock)((0, _data.useDispatch)(_store.store));
  (0, _element.useEffect)(() => {
    __experimentalUpdateSettings({
      ...settings,
      __internalIsInitialized: true
    }, {
      stripExperimentalSettings,
      reset: true
    });
  }, [settings, stripExperimentalSettings, __experimentalUpdateSettings]);

  // Syncs the entity provider with changes in the block-editor store.
  (0, _useBlockSync.default)(props);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.SlotFillProvider, {
    passthrough: true,
    children: [!settings?.__unstableIsPreviewMode && /*#__PURE__*/(0, _jsxRuntime.jsx)(_keyboardShortcuts.default.Register, {}), /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockRefsProvider.BlockRefsProvider, {
      children: children
    })]
  });
});
const BlockEditorProvider = props => {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(ExperimentalBlockEditorProvider, {
    ...props,
    stripExperimentalSettings: true,
    children: props.children
  });
};
exports.BlockEditorProvider = BlockEditorProvider;
var _default = exports.default = BlockEditorProvider;
