// @ariakit/core/cjs/__chunks/H6ZQSQQL.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _N667HVEDcjs = require('./N667HVED.cjs');





var _3QF3ODTRcjs = require('./3QF3ODTR.cjs');


var _BTUHABP4cjs = require('./BTUHABP4.cjs');




var _3IVNQNIVcjs = require('./3IVNQNIV.cjs');

// src/popover/popover-store.ts
function createPopoverStore(_a = {}) {
  var _b = _a, {
    popover: otherPopover
  } = _b, props = _3IVNQNIVcjs.__objRest.call(void 0, _b, [
    "popover"
  ]);
  const store = _3QF3ODTRcjs.mergeStore.call(void 0, 
    props.store,
    _3QF3ODTRcjs.omit.call(void 0, otherPopover, [
      "arrowElement",
      "anchorElement",
      "contentElement",
      "popoverElement",
      "disclosureElement"
    ])
  );
  _3QF3ODTRcjs.throwOnConflictingProps.call(void 0, props, store);
  const syncState = store == null ? void 0 : store.getState();
  const dialog = _N667HVEDcjs.createDialogStore.call(void 0, _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, props), { store }));
  const placement = _BTUHABP4cjs.defaultValue.call(void 0, 
    props.placement,
    syncState == null ? void 0 : syncState.placement,
    "bottom"
  );
  const initialState = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, dialog.getState()), {
    placement,
    currentPlacement: placement,
    anchorElement: _BTUHABP4cjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.anchorElement, null),
    popoverElement: _BTUHABP4cjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.popoverElement, null),
    arrowElement: _BTUHABP4cjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.arrowElement, null),
    rendered: Symbol("rendered")
  });
  const popover = _3QF3ODTRcjs.createStore.call(void 0, initialState, dialog, store);
  return _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, dialog), popover), {
    setAnchorElement: (element) => popover.setState("anchorElement", element),
    setPopoverElement: (element) => popover.setState("popoverElement", element),
    setArrowElement: (element) => popover.setState("arrowElement", element),
    render: () => popover.setState("rendered", Symbol("rendered"))
  });
}



exports.createPopoverStore = createPopoverStore;
