// @ariakit/core/cjs/__chunks/QLWAGBAQ.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";








var _3QF3ODTRcjs = require('./3QF3ODTR.cjs');


var _BTUHABP4cjs = require('./BTUHABP4.cjs');



var _3IVNQNIVcjs = require('./3IVNQNIV.cjs');

// src/disclosure/disclosure-store.ts
function createDisclosureStore(props = {}) {
  const store = _3QF3ODTRcjs.mergeStore.call(void 0, 
    props.store,
    _3QF3ODTRcjs.omit.call(void 0, props.disclosure, ["contentElement", "disclosureElement"])
  );
  _3QF3ODTRcjs.throwOnConflictingProps.call(void 0, props, store);
  const syncState = store == null ? void 0 : store.getState();
  const open = _BTUHABP4cjs.defaultValue.call(void 0, 
    props.open,
    syncState == null ? void 0 : syncState.open,
    props.defaultOpen,
    false
  );
  const animated = _BTUHABP4cjs.defaultValue.call(void 0, props.animated, syncState == null ? void 0 : syncState.animated, false);
  const initialState = {
    open,
    animated,
    animating: !!animated && open,
    mounted: open,
    contentElement: _BTUHABP4cjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.contentElement, null),
    disclosureElement: _BTUHABP4cjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.disclosureElement, null)
  };
  const disclosure = _3QF3ODTRcjs.createStore.call(void 0, initialState, store);
  _3QF3ODTRcjs.setup.call(void 0, 
    disclosure,
    () => _3QF3ODTRcjs.sync.call(void 0, disclosure, ["animated", "animating"], (state) => {
      if (state.animated) return;
      disclosure.setState("animating", false);
    })
  );
  _3QF3ODTRcjs.setup.call(void 0, 
    disclosure,
    () => _3QF3ODTRcjs.subscribe.call(void 0, disclosure, ["open"], () => {
      if (!disclosure.getState().animated) return;
      disclosure.setState("animating", true);
    })
  );
  _3QF3ODTRcjs.setup.call(void 0, 
    disclosure,
    () => _3QF3ODTRcjs.sync.call(void 0, disclosure, ["open", "animating"], (state) => {
      disclosure.setState("mounted", state.open || state.animating);
    })
  );
  return _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, disclosure), {
    disclosure: props.disclosure,
    setOpen: (value) => disclosure.setState("open", value),
    show: () => disclosure.setState("open", true),
    hide: () => disclosure.setState("open", false),
    toggle: () => disclosure.setState("open", (open2) => !open2),
    stopAnimation: () => disclosure.setState("animating", false),
    setContentElement: (value) => disclosure.setState("contentElement", value),
    setDisclosureElement: (value) => disclosure.setState("disclosureElement", value)
  });
}



exports.createDisclosureStore = createDisclosureStore;
