// @ariakit/core/cjs/__chunks/XIZWLBI2.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _H6ZQSQQLcjs = require('./H6ZQSQQL.cjs');


var _3QF3ODTRcjs = require('./3QF3ODTR.cjs');


var _BTUHABP4cjs = require('./BTUHABP4.cjs');



var _3IVNQNIVcjs = require('./3IVNQNIV.cjs');

// src/hovercard/hovercard-store.ts
function createHovercardStore(props = {}) {
  var _a;
  const syncState = (_a = props.store) == null ? void 0 : _a.getState();
  const popover = _H6ZQSQQLcjs.createPopoverStore.call(void 0, _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, props), {
    placement: _BTUHABP4cjs.defaultValue.call(void 0, 
      props.placement,
      syncState == null ? void 0 : syncState.placement,
      "bottom"
    )
  }));
  const timeout = _BTUHABP4cjs.defaultValue.call(void 0, props.timeout, syncState == null ? void 0 : syncState.timeout, 500);
  const initialState = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, popover.getState()), {
    timeout,
    showTimeout: _BTUHABP4cjs.defaultValue.call(void 0, props.showTimeout, syncState == null ? void 0 : syncState.showTimeout),
    hideTimeout: _BTUHABP4cjs.defaultValue.call(void 0, props.hideTimeout, syncState == null ? void 0 : syncState.hideTimeout),
    autoFocusOnShow: _BTUHABP4cjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.autoFocusOnShow, false)
  });
  const hovercard = _3QF3ODTRcjs.createStore.call(void 0, initialState, popover, props.store);
  return _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, popover), hovercard), {
    setAutoFocusOnShow: (value) => hovercard.setState("autoFocusOnShow", value)
  });
}



exports.createHovercardStore = createHovercardStore;
