// @ariakit/core/cjs/utils/dom.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



















var _7WYII4ANcjs = require('../__chunks/7WYII4AN.cjs');
require('../__chunks/3IVNQNIV.cjs');



















exports.canUseDOM = _7WYII4ANcjs.canUseDOM; exports.contains = _7WYII4ANcjs.contains; exports.getActiveElement = _7WYII4ANcjs.getActiveElement; exports.getDocument = _7WYII4ANcjs.getDocument; exports.getPopupItemRole = _7WYII4ANcjs.getPopupItemRole; exports.getPopupRole = _7WYII4ANcjs.getPopupRole; exports.getScrollingElement = _7WYII4ANcjs.getScrollingElement; exports.getTextboxSelection = _7WYII4ANcjs.getTextboxSelection; exports.getTextboxValue = _7WYII4ANcjs.getTextboxValue; exports.getWindow = _7WYII4ANcjs.getWindow; exports.isButton = _7WYII4ANcjs.isButton; exports.isFrame = _7WYII4ANcjs.isFrame; exports.isPartiallyHidden = _7WYII4ANcjs.isPartiallyHidden; exports.isTextField = _7WYII4ANcjs.isTextField; exports.isTextbox = _7WYII4ANcjs.isTextbox; exports.isVisible = _7WYII4ANcjs.isVisible; exports.scrollIntoViewIfNeeded = _7WYII4ANcjs.scrollIntoViewIfNeeded; exports.setSelectionRange = _7WYII4ANcjs.setSelectionRange;
