// @wordpress/dataviews/build/components/dataviews-bulk-actions/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BulkActions;
exports.useHasAPossibleBulkAction = useHasAPossibleBulkAction;
exports.useSomeItemHasAPossibleBulkAction = useSomeItemHasAPossibleBulkAction;
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _element = require("@wordpress/element");
var _data = require("@wordpress/data");
var _dataviewsContext = _interopRequireDefault(require("../dataviews-context"));
var _constants = require("../../constants");
var _lockUnlock = require("../../lock-unlock");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  DropdownMenuV2: DropdownMenu,
  DropdownMenuGroupV2: DropdownMenuGroup,
  DropdownMenuItemV2: DropdownMenuItem,
  DropdownMenuSeparatorV2: DropdownMenuSeparator
} = (0, _lockUnlock.unlock)(_components.privateApis);
function useHasAPossibleBulkAction(actions, item) {
  return (0, _element.useMemo)(() => {
    return actions.some(action => {
      return action.supportsBulk && (!action.isEligible || action.isEligible(item));
    });
  }, [actions, item]);
}
function useSomeItemHasAPossibleBulkAction(actions, data) {
  return (0, _element.useMemo)(() => {
    return data.some(item => {
      return actions.some(action => {
        return action.supportsBulk && (!action.isEligible || action.isEligible(item));
      });
    });
  }, [actions, data]);
}
function ActionWithModal({
  action,
  selectedItems,
  setActionWithModal,
  onMenuOpenChange
}) {
  const eligibleItems = (0, _element.useMemo)(() => {
    return selectedItems.filter(item => !action.isEligible || action.isEligible(item));
  }, [action, selectedItems]);
  const {
    RenderModal,
    hideModalHeader
  } = action;
  const onCloseModal = (0, _element.useCallback)(() => {
    setActionWithModal(undefined);
  }, [setActionWithModal]);
  if (!eligibleItems.length) {
    return null;
  }
  const label = typeof action.label === 'string' ? action.label : action.label(selectedItems);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Modal, {
    title: !hideModalHeader ? label : undefined,
    __experimentalHideHeader: !!hideModalHeader,
    onRequestClose: onCloseModal,
    overlayClassName: "dataviews-bulk-actions__modal",
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(RenderModal, {
      items: eligibleItems,
      closeModal: onCloseModal,
      onActionPerformed: () => onMenuOpenChange(false)
    })
  });
}
function BulkActionItem({
  action,
  selectedItems,
  setActionWithModal
}) {
  const registry = (0, _data.useRegistry)();
  const eligibleItems = (0, _element.useMemo)(() => {
    return selectedItems.filter(item => !action.isEligible || action.isEligible(item));
  }, [action, selectedItems]);
  const shouldShowModal = ('RenderModal' in action);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItem, {
    hideOnClick: !shouldShowModal,
    onClick: async () => {
      if (shouldShowModal) {
        setActionWithModal(action);
      } else {
        action.callback(eligibleItems, {
          registry
        });
      }
    },
    suffix: eligibleItems.length,
    children: action.label
  }, action.id);
}
function ActionsMenuGroup({
  actions,
  selectedItems,
  setActionWithModal
}) {
  const elligibleActions = (0, _element.useMemo)(() => {
    return actions.filter(action => {
      return selectedItems.some(item => !action.isEligible || action.isEligible(item));
    });
  }, [actions, selectedItems]);
  if (!elligibleActions.length) {
    return null;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuGroup, {
      children: elligibleActions.map(action => /*#__PURE__*/(0, _jsxRuntime.jsx)(BulkActionItem, {
        action: action,
        selectedItems: selectedItems,
        setActionWithModal: setActionWithModal
      }, action.id))
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuSeparator, {})]
  });
}
function _BulkActions() {
  const {
    data,
    actions = [],
    selection,
    onChangeSelection,
    getItemId
  } = (0, _element.useContext)(_dataviewsContext.default);
  const bulkActions = (0, _element.useMemo)(() => actions.filter(action => action.supportsBulk), [actions]);
  const [isMenuOpen, onMenuOpenChange] = (0, _element.useState)(false);
  const [actionWithModal, setActionWithModal] = (0, _element.useState)();
  const selectableItems = (0, _element.useMemo)(() => {
    return data.filter(item => {
      return bulkActions.some(action => !action.isEligible || action.isEligible(item));
    });
  }, [data, bulkActions]);
  const numberSelectableItems = selectableItems.length;
  const selectedItems = (0, _element.useMemo)(() => {
    return data.filter(item => selection.includes(getItemId(item)) && selectableItems.includes(item));
  }, [selection, data, getItemId, selectableItems]);
  const areAllSelected = selectedItems.length === numberSelectableItems;
  if (bulkActions.length === 0) {
    return null;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(DropdownMenu, {
      open: isMenuOpen,
      onOpenChange: onMenuOpenChange,
      label: (0, _i18n.__)('Bulk actions'),
      style: {
        minWidth: '240px'
      },
      trigger: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
        className: "dataviews-bulk-actions__edit-button",
        __next40pxDefaultSize: true,
        variant: "tertiary",
        size: "compact",
        children: selectedItems.length ? (0, _i18n.sprintf)( /* translators: %d: Number of items. */
        (0, _i18n._n)('Edit %d item', 'Edit %d items', selectedItems.length), selectedItems.length) : (0, _i18n.__)('Bulk edit')
      }),
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(ActionsMenuGroup, {
        actions: bulkActions,
        setActionWithModal: setActionWithModal,
        selectedItems: selectedItems
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(DropdownMenuGroup, {
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItem, {
          disabled: areAllSelected,
          hideOnClick: false,
          onClick: () => {
            onChangeSelection(selectableItems.map(item => getItemId(item)));
          },
          suffix: numberSelectableItems,
          children: (0, _i18n.__)('Select all')
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItem, {
          disabled: selection.length === 0,
          hideOnClick: false,
          onClick: () => {
            onChangeSelection([]);
          },
          children: (0, _i18n.__)('Deselect')
        })]
      })]
    }), actionWithModal && /*#__PURE__*/(0, _jsxRuntime.jsx)(ActionWithModal, {
      action: actionWithModal,
      selectedItems: selectedItems,
      setActionWithModal: setActionWithModal,
      onMenuOpenChange: onMenuOpenChange
    })]
  });
}
function BulkActions() {
  const {
    data,
    actions = [],
    view
  } = (0, _element.useContext)(_dataviewsContext.default);
  const hasPossibleBulkAction = useSomeItemHasAPossibleBulkAction(actions, data);
  if (![_constants.LAYOUT_TABLE, _constants.LAYOUT_GRID].includes(view.type) || !hasPossibleBulkAction) {
    return null;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_BulkActions, {});
}
