// @wordpress/dataviews/build-module/dataviews-layouts/table/column-header-menu.js
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { arrowLeft, arrowRight, unseen, funnel } from '@wordpress/icons';
import { Button, Icon, privateApis as componentsPrivateApis } from '@wordpress/components';
import { forwardRef, Children, Fragment } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { unlock } from '../../lock-unlock';
import { sanitizeOperators } from '../../utils';
import { SORTING_DIRECTIONS, sortArrows, sortLabels } from '../../constants';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
const {
  DropdownMenuV2: DropdownMenu,
  DropdownMenuGroupV2: DropdownMenuGroup,
  DropdownMenuItemV2: DropdownMenuItem,
  DropdownMenuRadioItemV2: DropdownMenuRadioItem,
  DropdownMenuItemLabelV2: DropdownMenuItemLabel,
  DropdownMenuSeparatorV2: DropdownMenuSeparator
} = unlock(componentsPrivateApis);
function WithDropDownMenuSeparators({
  children
}) {
  return Children.toArray(children).filter(Boolean).map((child, i) => /*#__PURE__*/_jsxs(Fragment, {
    children: [i > 0 && /*#__PURE__*/_jsx(DropdownMenuSeparator, {}), child]
  }, i));
}
const _HeaderMenu = forwardRef(function HeaderMenu({
  fieldId,
  view,
  fields,
  onChangeView,
  onHide,
  setOpenedFilter
}, ref) {
  const combinedField = view.layout?.combinedFields?.find(f => f.id === fieldId);
  const index = view.fields?.indexOf(fieldId);
  if (!!combinedField) {
    return combinedField.label;
  }
  const field = fields.find(f => f.id === fieldId);
  if (!field) {
    return null;
  }
  const isHidable = field.enableHiding !== false;
  const isSortable = field.enableSorting !== false;
  const isSorted = view.sort?.field === field.id;
  const operators = sanitizeOperators(field);
  // Filter can be added:
  // 1. If the field is not already part of a view's filters.
  // 2. If the field meets the type and operator requirements.
  // 3. If it's not primary. If it is, it should be already visible.
  const canAddFilter = !view.filters?.some(_filter => field.id === _filter.field) && !!field.elements?.length && !!operators.length && !field.filterBy?.isPrimary;
  if (!isSortable && !isHidable && !canAddFilter) {
    return field.label;
  }
  return /*#__PURE__*/_jsx(DropdownMenu, {
    align: "start",
    trigger: /*#__PURE__*/_jsxs(Button, {
      size: "compact",
      className: "dataviews-view-table-header-button",
      ref: ref,
      variant: "tertiary",
      children: [field.label, view.sort && isSorted && /*#__PURE__*/_jsx("span", {
        "aria-hidden": "true",
        children: sortArrows[view.sort.direction]
      })]
    }),
    style: {
      minWidth: '240px'
    },
    children: /*#__PURE__*/_jsxs(WithDropDownMenuSeparators, {
      children: [isSortable && /*#__PURE__*/_jsx(DropdownMenuGroup, {
        children: SORTING_DIRECTIONS.map(direction => {
          const isChecked = view.sort && isSorted && view.sort.direction === direction;
          const value = `${field.id}-${direction}`;
          return /*#__PURE__*/_jsx(DropdownMenuRadioItem, {
            // All sorting radio items share the same name, so that
            // selecting a sorting option automatically deselects the
            // previously selected one, even if it is displayed in
            // another submenu. The field and direction are passed via
            // the `value` prop.
            name: "view-table-sorting",
            value: value,
            checked: isChecked,
            onChange: () => {
              onChangeView({
                ...view,
                sort: {
                  field: field.id,
                  direction
                }
              });
            },
            children: /*#__PURE__*/_jsx(DropdownMenuItemLabel, {
              children: sortLabels[direction]
            })
          }, value);
        })
      }), canAddFilter && /*#__PURE__*/_jsx(DropdownMenuGroup, {
        children: /*#__PURE__*/_jsx(DropdownMenuItem, {
          prefix: /*#__PURE__*/_jsx(Icon, {
            icon: funnel
          }),
          onClick: () => {
            setOpenedFilter(field.id);
            onChangeView({
              ...view,
              page: 1,
              filters: [...(view.filters || []), {
                field: field.id,
                value: undefined,
                operator: operators[0]
              }]
            });
          },
          children: /*#__PURE__*/_jsx(DropdownMenuItemLabel, {
            children: __('Add filter')
          })
        })
      }), /*#__PURE__*/_jsxs(DropdownMenuGroup, {
        children: [/*#__PURE__*/_jsx(DropdownMenuItem, {
          prefix: /*#__PURE__*/_jsx(Icon, {
            icon: arrowLeft
          }),
          disabled: index < 1,
          onClick: () => {
            var _view$fields$slice;
            if (!view.fields || index < 1) {
              return;
            }
            onChangeView({
              ...view,
              fields: [...((_view$fields$slice = view.fields.slice(0, index - 1)) !== null && _view$fields$slice !== void 0 ? _view$fields$slice : []), field.id, view.fields[index - 1], ...view.fields.slice(index + 1)]
            });
          },
          children: /*#__PURE__*/_jsx(DropdownMenuItemLabel, {
            children: __('Move left')
          })
        }), /*#__PURE__*/_jsx(DropdownMenuItem, {
          prefix: /*#__PURE__*/_jsx(Icon, {
            icon: arrowRight
          }),
          disabled: !view.fields || index >= view.fields.length - 1,
          onClick: () => {
            var _view$fields$slice2;
            if (!view.fields || index >= view.fields.length - 1) {
              return;
            }
            onChangeView({
              ...view,
              fields: [...((_view$fields$slice2 = view.fields.slice(0, index)) !== null && _view$fields$slice2 !== void 0 ? _view$fields$slice2 : []), view.fields[index + 1], field.id, ...view.fields.slice(index + 2)]
            });
          },
          children: /*#__PURE__*/_jsx(DropdownMenuItemLabel, {
            children: __('Move right')
          })
        }), isHidable && /*#__PURE__*/_jsx(DropdownMenuItem, {
          prefix: /*#__PURE__*/_jsx(Icon, {
            icon: unseen
          }),
          onClick: () => {
            const viewFields = view.fields || fields.map(f => f.id);
            onHide(field);
            onChangeView({
              ...view,
              fields: viewFields.filter(id => id !== field.id)
            });
          },
          children: /*#__PURE__*/_jsx(DropdownMenuItemLabel, {
            children: __('Hide column')
          })
        })]
      })]
    })
  });
});

// @ts-expect-error Lift the `Item` type argument through the forwardRef.
const ColumnHeaderMenu = _HeaderMenu;
export default ColumnHeaderMenu;
