// @wordpress/dataviews/build/field-types/integer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _i18n = require("@wordpress/i18n");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function sort(a, b, direction) {
  return direction === 'asc' ? a - b : b - a;
}
function isValid(value, context) {
  // TODO: this implicitely means the value is required.
  if (value === '') {
    return false;
  }
  if (!Number.isInteger(Number(value))) {
    return false;
  }
  if (context?.elements) {
    const validValues = context?.elements.map(f => f.value);
    if (!validValues.includes(Number(value))) {
      return false;
    }
  }
  return true;
}
function Edit({
  data,
  field,
  onChange,
  hideLabelFromVision
}) {
  var _field$getValue;
  const {
    id,
    label,
    description
  } = field;
  const value = (_field$getValue = field.getValue({
    item: data
  })) !== null && _field$getValue !== void 0 ? _field$getValue : '';
  const onChangeControl = (0, _element.useCallback)(newValue => onChange(prevItem => ({
    ...prevItem,
    [id]: newValue
  })), [id, onChange]);
  if (field.elements) {
    const elements = [
    /*
     * Value can be undefined when:
     *
     * - the field is not required
     * - in bulk editing
     *
     */
    {
      label: (0, _i18n.__)('Select item'),
      value: ''
    }, ...field.elements];
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.SelectControl, {
      label: label,
      value: value,
      options: elements,
      onChange: onChangeControl,
      __next40pxDefaultSize: true,
      __nextHasNoMarginBottom: true,
      hideLabelFromVision: hideLabelFromVision
    });
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalNumberControl, {
    label: label,
    help: description,
    value: value,
    onChange: onChangeControl,
    __next40pxDefaultSize: true,
    hideLabelFromVision: hideLabelFromVision
  });
}
var _default = exports.default = {
  sort,
  isValid,
  Edit
};
