// @wordpress/dataviews/build/field-types/text.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _i18n = require("@wordpress/i18n");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function sort(valueA, valueB, direction) {
  return direction === 'asc' ? valueA.localeCompare(valueB) : valueB.localeCompare(valueA);
}
function isValid(value, context) {
  if (context?.elements) {
    const validValues = context?.elements?.map(f => f.value);
    if (!validValues.includes(value)) {
      return false;
    }
  }
  return true;
}
function Edit({
  data,
  field,
  onChange,
  hideLabelFromVision
}) {
  const {
    id,
    label,
    placeholder
  } = field;
  const value = field.getValue({
    item: data
  });
  const onChangeControl = (0, _element.useCallback)(newValue => onChange(prevItem => ({
    ...prevItem,
    [id]: newValue
  })), [id, onChange]);
  if (field.elements) {
    const elements = [
    /*
     * Value can be undefined when:
     *
     * - the field is not required
     * - in bulk editing
     *
     */
    {
      label: (0, _i18n.__)('Select item'),
      value: ''
    }, ...field.elements];
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.SelectControl, {
      label: label,
      value: value,
      options: elements,
      onChange: onChangeControl,
      __next40pxDefaultSize: true,
      __nextHasNoMarginBottom: true,
      hideLabelFromVision: hideLabelFromVision
    });
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.TextControl, {
    label: label,
    placeholder: placeholder,
    value: value !== null && value !== void 0 ? value : '',
    onChange: onChangeControl,
    __next40pxDefaultSize: true,
    __nextHasNoMarginBottom: true,
    hideLabelFromVision: hideLabelFromVision
  });
}
var _default = exports.default = {
  sort,
  isValid,
  Edit
};
