// @wordpress/dataviews/build/dataviews-layouts/grid/density-picker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DensityPicker;
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _compose = require("@wordpress/compose");
var _icons = require("@wordpress/icons");
var _element = require("@wordpress/element");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

const viewportBreaks = {
  xhuge: {
    min: 3,
    max: 6,
    default: 5
  },
  huge: {
    min: 2,
    max: 4,
    default: 4
  },
  xlarge: {
    min: 2,
    max: 3,
    default: 3
  },
  large: {
    min: 1,
    max: 2,
    default: 2
  },
  mobile: {
    min: 1,
    max: 2,
    default: 2
  }
};
function useViewPortBreakpoint() {
  const isXHuge = (0, _compose.useViewportMatch)('xhuge', '>=');
  const isHuge = (0, _compose.useViewportMatch)('huge', '>=');
  const isXlarge = (0, _compose.useViewportMatch)('xlarge', '>=');
  const isLarge = (0, _compose.useViewportMatch)('large', '>=');
  const isMobile = (0, _compose.useViewportMatch)('mobile', '>=');
  if (isXHuge) {
    return 'xhuge';
  }
  if (isHuge) {
    return 'huge';
  }
  if (isXlarge) {
    return 'xlarge';
  }
  if (isLarge) {
    return 'large';
  }
  if (isMobile) {
    return 'mobile';
  }
  return null;
}

// Value is number from 0 to 100 representing how big an item is in the grid
// 100 being the biggest and 0 being the smallest.
// The size is relative to the viewport size, if one a given viewport the
// number of allowed items in a grid is 3 to 6 a 0 ( the smallest ) will mean that the grid will
// have 6 items in a row, a 100 ( the biggest ) will mean that the grid will have 3 items in a row.
// A value of 75 will mean that the grid will have 4 items in a row.
function getRangeValue(density, breakValues) {
  const inverseDensity = breakValues.max - density;
  const max = breakValues.max - breakValues.min;
  return Math.round(inverseDensity * 100 / max);
}
function DensityPicker({
  density,
  setDensity
}) {
  const viewport = useViewPortBreakpoint();
  (0, _element.useEffect)(() => {
    setDensity(_density => {
      if (!viewport || !_density) {
        return 0;
      }
      const breakValues = viewportBreaks[viewport];
      if (_density < breakValues.min) {
        return breakValues.min;
      }
      if (_density > breakValues.max) {
        return breakValues.max;
      }
      return _density;
    });
  }, [setDensity, viewport]);
  if (!viewport) {
    return null;
  }
  const breakValues = viewportBreaks[viewport];
  const densityToUse = density || breakValues.default;
  const rangeValue = getRangeValue(densityToUse, breakValues);
  const step = 100 / (breakValues.max - breakValues.min + 1);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
      size: "compact",
      icon: _icons.reset,
      disabled: rangeValue <= 0,
      accessibleWhenDisabled: true,
      label: (0, _i18n.__)('Decrease size'),
      onClick: () => {
        setDensity(densityToUse + 1);
      }
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.RangeControl, {
      __nextHasNoMarginBottom: true,
      showTooltip: false,
      className: "dataviews-density-picker__range-control",
      label: (0, _i18n.__)('Item size'),
      hideLabelFromVision: true,
      value: rangeValue,
      min: 0,
      max: 100,
      withInputField: false,
      onChange: (value = 0) => {
        const inverseValue = 100 - value;
        setDensity(Math.round(inverseValue * (breakValues.max - breakValues.min) / 100 + breakValues.min));
      },
      step: step
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
      size: "compact",
      icon: _icons.plus,
      disabled: rangeValue >= 100,
      accessibleWhenDisabled: true,
      label: (0, _i18n.__)('Increase size'),
      onClick: () => {
        setDensity(densityToUse - 1);
      }
    })]
  });
}
