// @wordpress/dataviews/build/components/dataviews-view-config/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _element = require("@wordpress/element");
var _icons = require("@wordpress/icons");
var _warning = _interopRequireDefault(require("@wordpress/warning"));
var _constants = require("../../constants");
var _dataviewsLayouts = require("../../dataviews-layouts");
var _dataviewsContext = _interopRequireDefault(require("../dataviews-context"));
var _lockUnlock = require("../../lock-unlock");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  DropdownMenuV2: DropdownMenu,
  DropdownMenuRadioItemV2: DropdownMenuRadioItem,
  DropdownMenuItemLabelV2: DropdownMenuItemLabel
} = (0, _lockUnlock.unlock)(_components.privateApis);
function ViewTypeMenu({
  defaultLayouts = {
    list: {},
    grid: {},
    table: {}
  }
}) {
  const {
    view,
    onChangeView
  } = (0, _element.useContext)(_dataviewsContext.default);
  const availableLayouts = Object.keys(defaultLayouts);
  if (availableLayouts.length <= 1) {
    return null;
  }
  const activeView = _dataviewsLayouts.VIEW_LAYOUTS.find(v => view.type === v.type);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenu, {
    trigger: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
      size: "compact",
      icon: activeView?.icon,
      label: (0, _i18n.__)('Layout')
    }),
    children: availableLayouts.map(layout => {
      const config = _dataviewsLayouts.VIEW_LAYOUTS.find(v => v.type === layout);
      if (!config) {
        return null;
      }
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuRadioItem, {
        value: layout,
        name: "view-actions-available-view",
        checked: layout === view.type,
        hideOnClick: true,
        onChange: e => {
          switch (e.target.value) {
            case 'list':
            case 'grid':
            case 'table':
              return onChangeView({
                ...view,
                type: e.target.value,
                ...defaultLayouts[e.target.value]
              });
          }
          globalThis.SCRIPT_DEBUG === true ? (0, _warning.default)('Invalid dataview') : void 0;
        },
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItemLabel, {
          children: config.label
        })
      }, layout);
    })
  });
}
function SortFieldControl() {
  const {
    view,
    fields,
    onChangeView
  } = (0, _element.useContext)(_dataviewsContext.default);
  const orderOptions = (0, _element.useMemo)(() => {
    const sortableFields = fields.filter(field => field.enableSorting !== false);
    return sortableFields.map(field => {
      return {
        label: field.label,
        value: field.id
      };
    });
  }, [fields]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.SelectControl, {
    __nextHasNoMarginBottom: true,
    __next40pxDefaultSize: true,
    label: (0, _i18n.__)('Sort by'),
    value: view.sort?.field,
    options: orderOptions,
    onChange: value => {
      onChangeView({
        ...view,
        sort: {
          direction: view?.sort?.direction || 'desc',
          field: value
        }
      });
    }
  });
}
function SortDirectionControl() {
  const {
    view,
    onChangeView
  } = (0, _element.useContext)(_dataviewsContext.default);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalToggleGroupControl, {
    className: "dataviews-view-config__sort-direction",
    __nextHasNoMarginBottom: true,
    __next40pxDefaultSize: true,
    isBlock: true,
    label: (0, _i18n.__)('Order'),
    value: view.sort?.direction || 'desc',
    disabled: !view?.sort?.field,
    onChange: newDirection => {
      if (!view?.sort?.field) {
        return;
      }
      if (newDirection === 'asc' || newDirection === 'desc') {
        onChangeView({
          ...view,
          sort: {
            direction: newDirection,
            field: view.sort.field
          }
        });
        return;
      }
      globalThis.SCRIPT_DEBUG === true ? (0, _warning.default)('Invalid direction') : void 0;
    },
    children: _constants.SORTING_DIRECTIONS.map(direction => {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalToggleGroupControlOptionIcon, {
        value: direction,
        icon: _constants.sortIcons[direction],
        label: _constants.sortLabels[direction]
      }, direction);
    })
  });
}
const PAGE_SIZE_VALUES = [10, 20, 50, 100];
function ItemsPerPageControl() {
  const {
    view,
    onChangeView
  } = (0, _element.useContext)(_dataviewsContext.default);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalToggleGroupControl, {
    __nextHasNoMarginBottom: true,
    __next40pxDefaultSize: true,
    isBlock: true,
    label: (0, _i18n.__)('Items per page'),
    value: view.perPage || 10,
    disabled: !view?.sort?.field,
    onChange: newItemsPerPage => {
      const newItemsPerPageNumber = typeof newItemsPerPage === 'number' || newItemsPerPage === undefined ? newItemsPerPage : parseInt(newItemsPerPage, 10);
      onChangeView({
        ...view,
        perPage: newItemsPerPageNumber,
        page: 1
      });
    },
    children: PAGE_SIZE_VALUES.map(value => {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalToggleGroupControlOption, {
        value: value,
        label: value.toString()
      }, value);
    })
  });
}
function FieldControl() {
  const {
    view,
    fields,
    onChangeView
  } = (0, _element.useContext)(_dataviewsContext.default);
  const mandatoryFields = (0, _dataviewsLayouts.getMandatoryFields)(view);
  const hidableFields = fields.filter(field => field.enableHiding !== false && !mandatoryFields.includes(field.id));
  const viewFields = view.fields || fields.map(field => field.id);
  if (!hidableFields?.length) {
    return null;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalItemGroup, {
    isBordered: true,
    isSeparated: true,
    children: hidableFields?.map(field => {
      const isVisible = viewFields.includes(field.id);
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalItem, {
        children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
          expanded: true,
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
            children: field.label
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
            className: "'dataviews-view-config__field-control-button",
            size: "compact",
            onClick: () => onChangeView({
              ...view,
              fields: isVisible ? viewFields.filter(id => id !== field.id) : [...viewFields, field.id]
            }),
            icon: isVisible ? _icons.seen : _icons.unseen,
            label: isVisible ? (0, _i18n.__)('Hide field') : (0, _i18n.__)('Show field')
          })]
        })
      }, field.id);
    })
  });
}
function SettingsSection({
  title,
  description,
  children
}) {
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalGrid, {
    columns: 12,
    className: "dataviews-settings-section",
    gap: 4,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: "dataviews-settings-section__sidebar",
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalHeading, {
        level: 2,
        className: "dataviews-settings-section__title",
        children: title
      }), description && /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalText, {
        variant: "muted",
        className: "dataviews-settings-section__description",
        children: description
      })]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalGrid, {
      columns: 8,
      gap: 4,
      className: "dataviews-settings-section__content",
      children: children
    })]
  });
}
function DataviewsViewConfigContent() {
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalVStack, {
    className: "dataviews-view-config",
    spacing: 6,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(SettingsSection, {
      title: (0, _i18n.__)('Appearance'),
      children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
        expanded: true,
        className: "is-divided-in-two",
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(SortFieldControl, {}), /*#__PURE__*/(0, _jsxRuntime.jsx)(SortDirectionControl, {})]
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(ItemsPerPageControl, {})]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(SettingsSection, {
      title: (0, _i18n.__)('Properties'),
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(FieldControl, {})
    })]
  });
}
function _DataViewsViewConfig({
  defaultLayouts = {
    list: {},
    grid: {},
    table: {}
  }
}) {
  const [isShowingViewPopover, setIsShowingViewPopover] = (0, _element.useState)(false);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(ViewTypeMenu, {
      defaultLayouts: defaultLayouts
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
        size: "compact",
        icon: _icons.cog,
        label: (0, _i18n._x)('View options', 'View is used as a noun'),
        onClick: () => setIsShowingViewPopover(true)
      }), isShowingViewPopover && /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Popover, {
        placement: "bottom-end",
        onClose: () => {
          setIsShowingViewPopover(false);
        },
        focusOnMount: true,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DataviewsViewConfigContent, {})
      })]
    })]
  });
}
const DataViewsViewConfig = (0, _element.memo)(_DataViewsViewConfig);
var _default = exports.default = DataViewsViewConfig;
