// @wordpress/block-editor/build-module/components/inserter/category-tabs/index.js
/**
 * WordPress dependencies
 */
import { usePrevious, useReducedMotion } from '@wordpress/compose';
import { isRTL } from '@wordpress/i18n';
import { __experimentalHStack as HStack, FlexBlock, privateApis as componentsPrivateApis, __unstableMotion as motion } from '@wordpress/components';
import { Icon, chevronRight, chevronLeft } from '@wordpress/icons';

/**
 * Internal dependencies
 */
import { unlock } from '../../../lock-unlock';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
const {
  Tabs
} = unlock(componentsPrivateApis);
function CategoryTabs({
  categories,
  selectedCategory,
  onSelectCategory,
  children
}) {
  // Copied from InterfaceSkeleton.
  const ANIMATION_DURATION = 0.25;
  const disableMotion = useReducedMotion();
  const defaultTransition = {
    type: 'tween',
    duration: disableMotion ? 0 : ANIMATION_DURATION,
    ease: [0.6, 0, 0.4, 1]
  };
  const previousSelectedCategory = usePrevious(selectedCategory);
  return /*#__PURE__*/_jsxs(Tabs, {
    className: "block-editor-inserter__category-tabs",
    selectOnMove: false,
    selectedTabId: selectedCategory ? selectedCategory.name : null,
    orientation: "vertical",
    onSelect: categoryId => {
      // Pass the full category object
      onSelectCategory(categories.find(category => category.name === categoryId));
    },
    children: [/*#__PURE__*/_jsx(Tabs.TabList, {
      className: "block-editor-inserter__category-tablist",
      children: categories.map(category => /*#__PURE__*/_jsx(Tabs.Tab, {
        tabId: category.name,
        className: "block-editor-inserter__category-tab",
        "aria-label": category.label,
        "aria-current": category === selectedCategory ? 'true' : undefined,
        children: /*#__PURE__*/_jsxs(HStack, {
          children: [/*#__PURE__*/_jsx(FlexBlock, {
            children: category.label
          }), /*#__PURE__*/_jsx(Icon, {
            icon: isRTL() ? chevronLeft : chevronRight
          })]
        })
      }, category.name))
    }), categories.map(category => /*#__PURE__*/_jsx(Tabs.TabPanel, {
      tabId: category.name,
      focusable: false,
      children: /*#__PURE__*/_jsx(motion.div, {
        className: "block-editor-inserter__category-panel",
        initial: !previousSelectedCategory ? 'closed' : 'open',
        animate: "open",
        variants: {
          open: {
            transform: 'translateX( 0 )',
            transitionEnd: {
              zIndex: '1'
            }
          },
          closed: {
            transform: 'translateX( -100% )',
            zIndex: '-1'
          }
        },
        transition: defaultTransition,
        children: children
      })
    }, category.name))]
  });
}
export default CategoryTabs;
