// @ariakit/core/cjs/tab/tab-store.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _PW36G76Lcjs = require('../__chunks/PW36G76L.cjs');


var _CCI5VM7Ecjs = require('../__chunks/CCI5VM7E.cjs');
require('../__chunks/ULSPM3Y3.cjs');







var _3QF3ODTRcjs = require('../__chunks/3QF3ODTR.cjs');



var _BTUHABP4cjs = require('../__chunks/BTUHABP4.cjs');
require('../__chunks/7WYII4AN.cjs');




var _3IVNQNIVcjs = require('../__chunks/3IVNQNIV.cjs');

// src/tab/tab-store.ts
function createTabStore(_a = {}) {
  var _b = _a, {
    composite: parentComposite,
    combobox
  } = _b, props = _3IVNQNIVcjs.__objRest.call(void 0, _b, [
    "composite",
    "combobox"
  ]);
  const independentKeys = [
    "items",
    "renderedItems",
    "moves",
    "orientation",
    "virtualFocus",
    "includesBaseElement",
    "baseElement",
    "focusLoop",
    "focusShift",
    "focusWrap"
  ];
  const store = _3QF3ODTRcjs.mergeStore.call(void 0, 
    props.store,
    _3QF3ODTRcjs.omit.call(void 0, parentComposite, independentKeys),
    _3QF3ODTRcjs.omit.call(void 0, combobox, independentKeys)
  );
  const syncState = store == null ? void 0 : store.getState();
  const composite = _PW36G76Lcjs.createCompositeStore.call(void 0, _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, props), {
    store,
    // We need to explicitly set the default value of `includesBaseElement` to
    // `false` since we don't want the composite store to default it to `true`
    // when the activeId state is null, which could be the case when rendering
    // combobox with tab.
    includesBaseElement: _BTUHABP4cjs.defaultValue.call(void 0, 
      props.includesBaseElement,
      syncState == null ? void 0 : syncState.includesBaseElement,
      false
    ),
    orientation: _BTUHABP4cjs.defaultValue.call(void 0, 
      props.orientation,
      syncState == null ? void 0 : syncState.orientation,
      "horizontal"
    ),
    focusLoop: _BTUHABP4cjs.defaultValue.call(void 0, props.focusLoop, syncState == null ? void 0 : syncState.focusLoop, true)
  }));
  const panels = _CCI5VM7Ecjs.createCollectionStore.call(void 0, );
  const initialState = _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, composite.getState()), {
    selectedId: _BTUHABP4cjs.defaultValue.call(void 0, 
      props.selectedId,
      syncState == null ? void 0 : syncState.selectedId,
      props.defaultSelectedId
    ),
    selectOnMove: _BTUHABP4cjs.defaultValue.call(void 0, 
      props.selectOnMove,
      syncState == null ? void 0 : syncState.selectOnMove,
      true
    )
  });
  const tab = _3QF3ODTRcjs.createStore.call(void 0, initialState, composite, store);
  _3QF3ODTRcjs.setup.call(void 0, 
    tab,
    () => _3QF3ODTRcjs.sync.call(void 0, tab, ["moves"], () => {
      const { activeId, selectOnMove } = tab.getState();
      if (!selectOnMove) return;
      if (!activeId) return;
      const tabItem = composite.item(activeId);
      if (!tabItem) return;
      if (tabItem.dimmed) return;
      if (tabItem.disabled) return;
      tab.setState("selectedId", tabItem.id);
    })
  );
  _3QF3ODTRcjs.setup.call(void 0, 
    tab,
    () => _3QF3ODTRcjs.batch.call(void 0, tab, ["selectedId"], (state, prev) => {
      if (parentComposite && state.selectedId === prev.selectedId) return;
      tab.setState("activeId", state.selectedId);
    })
  );
  _3QF3ODTRcjs.setup.call(void 0, 
    tab,
    () => _3QF3ODTRcjs.sync.call(void 0, tab, ["selectedId", "renderedItems"], (state) => {
      if (state.selectedId !== void 0) return;
      const { activeId, renderedItems } = tab.getState();
      const tabItem = composite.item(activeId);
      if (tabItem && !tabItem.disabled && !tabItem.dimmed) {
        tab.setState("selectedId", tabItem.id);
      } else {
        const tabItem2 = renderedItems.find(
          (item) => !item.disabled && !item.dimmed
        );
        tab.setState("selectedId", tabItem2 == null ? void 0 : tabItem2.id);
      }
    })
  );
  _3QF3ODTRcjs.setup.call(void 0, 
    tab,
    () => _3QF3ODTRcjs.sync.call(void 0, tab, ["renderedItems"], (state) => {
      const tabs = state.renderedItems;
      if (!tabs.length) return;
      return _3QF3ODTRcjs.sync.call(void 0, panels, ["renderedItems"], (state2) => {
        const items = state2.renderedItems;
        const hasOrphanPanels = items.some((panel) => !panel.tabId);
        if (!hasOrphanPanels) return;
        items.forEach((panel, i) => {
          if (panel.tabId) return;
          const tabItem = tabs[i];
          if (!tabItem) return;
          panels.renderItem(_3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, panel), { tabId: tabItem.id }));
        });
      });
    })
  );
  let selectedIdFromSelectedValue = null;
  _3QF3ODTRcjs.setup.call(void 0, tab, () => {
    const backupSelectedId = () => {
      selectedIdFromSelectedValue = tab.getState().selectedId;
    };
    const restoreSelectedId = () => {
      tab.setState("selectedId", selectedIdFromSelectedValue);
    };
    if (parentComposite && "setSelectElement" in parentComposite) {
      return _BTUHABP4cjs.chain.call(void 0, 
        _3QF3ODTRcjs.sync.call(void 0, parentComposite, ["value"], backupSelectedId),
        _3QF3ODTRcjs.sync.call(void 0, parentComposite, ["open"], restoreSelectedId)
      );
    }
    if (!combobox) return;
    return _BTUHABP4cjs.chain.call(void 0, 
      _3QF3ODTRcjs.sync.call(void 0, combobox, ["selectedValue"], backupSelectedId),
      _3QF3ODTRcjs.sync.call(void 0, combobox, ["open"], restoreSelectedId)
    );
  });
  return _3IVNQNIVcjs.__spreadProps.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, _3IVNQNIVcjs.__spreadValues.call(void 0, {}, composite), tab), {
    panels,
    setSelectedId: (id) => tab.setState("selectedId", id),
    select: (id) => {
      tab.setState("selectedId", id);
      composite.move(id);
    }
  });
}


exports.createTabStore = createTabStore;
