// @wordpress/dataviews/build/dataviews-layouts/table/column-header-menu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _lockUnlock = require("../../lock-unlock");
var _utils = require("../../utils");
var _constants = require("../../constants");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  DropdownMenuV2: DropdownMenu,
  DropdownMenuGroupV2: DropdownMenuGroup,
  DropdownMenuItemV2: DropdownMenuItem,
  DropdownMenuRadioItemV2: DropdownMenuRadioItem,
  DropdownMenuItemLabelV2: DropdownMenuItemLabel,
  DropdownMenuSeparatorV2: DropdownMenuSeparator
} = (0, _lockUnlock.unlock)(_components.privateApis);
function WithDropDownMenuSeparators({
  children
}) {
  return _element.Children.toArray(children).filter(Boolean).map((child, i) => /*#__PURE__*/(0, _jsxRuntime.jsxs)(_element.Fragment, {
    children: [i > 0 && /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuSeparator, {}), child]
  }, i));
}
const _HeaderMenu = (0, _element.forwardRef)(function HeaderMenu({
  fieldId,
  view,
  fields,
  onChangeView,
  onHide,
  setOpenedFilter
}, ref) {
  const combinedField = view.layout?.combinedFields?.find(f => f.id === fieldId);
  const index = view.fields?.indexOf(fieldId);
  if (!!combinedField) {
    return combinedField.label;
  }
  const field = fields.find(f => f.id === fieldId);
  if (!field) {
    return null;
  }
  const isHidable = field.enableHiding !== false;
  const isSortable = field.enableSorting !== false;
  const isSorted = view.sort?.field === field.id;
  const operators = (0, _utils.sanitizeOperators)(field);
  // Filter can be added:
  // 1. If the field is not already part of a view's filters.
  // 2. If the field meets the type and operator requirements.
  // 3. If it's not primary. If it is, it should be already visible.
  const canAddFilter = !view.filters?.some(_filter => field.id === _filter.field) && !!field.elements?.length && !!operators.length && !field.filterBy?.isPrimary;
  if (!isSortable && !isHidable && !canAddFilter) {
    return field.label;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenu, {
    align: "start",
    trigger: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.Button, {
      size: "compact",
      className: "dataviews-view-table-header-button",
      ref: ref,
      variant: "tertiary",
      children: [field.label, view.sort && isSorted && /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
        "aria-hidden": "true",
        children: _constants.sortArrows[view.sort.direction]
      })]
    }),
    style: {
      minWidth: '240px'
    },
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(WithDropDownMenuSeparators, {
      children: [isSortable && /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuGroup, {
        children: _constants.SORTING_DIRECTIONS.map(direction => {
          const isChecked = view.sort && isSorted && view.sort.direction === direction;
          const value = `${field.id}-${direction}`;
          return /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuRadioItem, {
            // All sorting radio items share the same name, so that
            // selecting a sorting option automatically deselects the
            // previously selected one, even if it is displayed in
            // another submenu. The field and direction are passed via
            // the `value` prop.
            name: "view-table-sorting",
            value: value,
            checked: isChecked,
            onChange: () => {
              onChangeView({
                ...view,
                sort: {
                  field: field.id,
                  direction
                }
              });
            },
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItemLabel, {
              children: _constants.sortLabels[direction]
            })
          }, value);
        })
      }), canAddFilter && /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuGroup, {
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItem, {
          prefix: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Icon, {
            icon: _icons.funnel
          }),
          onClick: () => {
            setOpenedFilter(field.id);
            onChangeView({
              ...view,
              page: 1,
              filters: [...(view.filters || []), {
                field: field.id,
                value: undefined,
                operator: operators[0]
              }]
            });
          },
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItemLabel, {
            children: (0, _i18n.__)('Add filter')
          })
        })
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(DropdownMenuGroup, {
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItem, {
          prefix: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Icon, {
            icon: _icons.arrowLeft
          }),
          disabled: index < 1,
          onClick: () => {
            var _view$fields$slice;
            if (!view.fields || index < 1) {
              return;
            }
            onChangeView({
              ...view,
              fields: [...((_view$fields$slice = view.fields.slice(0, index - 1)) !== null && _view$fields$slice !== void 0 ? _view$fields$slice : []), field.id, view.fields[index - 1], ...view.fields.slice(index + 1)]
            });
          },
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItemLabel, {
            children: (0, _i18n.__)('Move left')
          })
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItem, {
          prefix: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Icon, {
            icon: _icons.arrowRight
          }),
          disabled: !view.fields || index >= view.fields.length - 1,
          onClick: () => {
            var _view$fields$slice2;
            if (!view.fields || index >= view.fields.length - 1) {
              return;
            }
            onChangeView({
              ...view,
              fields: [...((_view$fields$slice2 = view.fields.slice(0, index)) !== null && _view$fields$slice2 !== void 0 ? _view$fields$slice2 : []), view.fields[index + 1], field.id, ...view.fields.slice(index + 2)]
            });
          },
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItemLabel, {
            children: (0, _i18n.__)('Move right')
          })
        }), isHidable && /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItem, {
          prefix: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Icon, {
            icon: _icons.unseen
          }),
          onClick: () => {
            const viewFields = view.fields || fields.map(f => f.id);
            onHide(field);
            onChangeView({
              ...view,
              fields: viewFields.filter(id => id !== field.id)
            });
          },
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownMenuItemLabel, {
            children: (0, _i18n.__)('Hide column')
          })
        })]
      })]
    })
  });
});

// @ts-expect-error Lift the `Item` type argument through the forwardRef.
const ColumnHeaderMenu = _HeaderMenu;
var _default = exports.default = ColumnHeaderMenu;
