// @wordpress/dataviews/build/components/dataviews-pagination/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
var _dataviewsContext = _interopRequireDefault(require("../dataviews-context"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function DataViewsPagination() {
  var _view$page;
  const {
    view,
    onChangeView,
    paginationInfo: {
      totalItems = 0,
      totalPages
    }
  } = (0, _element.useContext)(_dataviewsContext.default);
  if (!totalItems || !totalPages) {
    return null;
  }
  const currentPage = (_view$page = view.page) !== null && _view$page !== void 0 ? _view$page : 1;
  return !!totalItems && totalPages !== 1 && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
    expanded: false,
    spacing: 6,
    justify: "end",
    className: "dataviews-pagination",
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalHStack, {
      justify: "flex-start",
      expanded: false,
      spacing: 2,
      className: "dataviews-pagination__page-selection",
      children: (0, _element.createInterpolateElement)((0, _i18n.sprintf)(
      // translators: %s: Total number of pages.
      (0, _i18n._x)('Page <CurrentPageControl /> of %s', 'paging'), totalPages), {
        CurrentPageControl: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.SelectControl, {
          "aria-label": (0, _i18n.__)('Current page'),
          value: view.page?.toString(),
          options: Array.from(Array(totalPages)).map((_, i) => {
            const page = i + 1;
            return {
              value: page.toString(),
              label: page.toString()
            };
          }),
          onChange: newValue => {
            onChangeView({
              ...view,
              page: +newValue
            });
          },
          size: "compact",
          __nextHasNoMarginBottom: true
        })
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
      expanded: false,
      spacing: 1,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
        onClick: () => onChangeView({
          ...view,
          page: currentPage - 1
        }),
        disabled: currentPage === 1,
        accessibleWhenDisabled: true,
        label: (0, _i18n.__)('Previous page'),
        icon: _icons.previous,
        showTooltip: true,
        size: "compact",
        tooltipPosition: "top"
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
        onClick: () => onChangeView({
          ...view,
          page: currentPage + 1
        }),
        disabled: currentPage >= totalPages,
        accessibleWhenDisabled: true,
        label: (0, _i18n.__)('Next page'),
        icon: _icons.next,
        showTooltip: true,
        size: "compact",
        tooltipPosition: "top"
      })]
    })]
  });
}
var _default = exports.default = (0, _element.memo)(DataViewsPagination);
