// @wordpress/patterns/build/components/pattern-overrides-controls.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _element = require("@wordpress/element");
var _blockEditor = require("@wordpress/block-editor");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _constants = require("../constants");
var _allowOverridesModal = require("./allow-overrides-modal");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function removeBindings(bindings) {
  let updatedBindings = {
    ...bindings
  };
  delete updatedBindings.__default;
  if (!Object.keys(updatedBindings).length) {
    updatedBindings = undefined;
  }
  return updatedBindings;
}
function addBindings(bindings) {
  return {
    ...bindings,
    __default: {
      source: _constants.PATTERN_OVERRIDES_BINDING_SOURCE
    }
  };
}
function PatternOverridesControls({
  attributes,
  setAttributes,
  name: blockName
}) {
  const controlId = (0, _element.useId)();
  const [showAllowOverridesModal, setShowAllowOverridesModal] = (0, _element.useState)(false);
  const [showDisallowOverridesModal, setShowDisallowOverridesModal] = (0, _element.useState)(false);
  const hasName = !!attributes.metadata?.name;
  const defaultBindings = attributes.metadata?.bindings?.__default;
  const hasOverrides = hasName && defaultBindings?.source === _constants.PATTERN_OVERRIDES_BINDING_SOURCE;
  const isConnectedToOtherSources = defaultBindings?.source && defaultBindings.source !== _constants.PATTERN_OVERRIDES_BINDING_SOURCE;
  function updateBindings(isChecked, customName) {
    const prevBindings = attributes?.metadata?.bindings;
    const updatedBindings = isChecked ? addBindings(prevBindings) : removeBindings(prevBindings);
    const updatedMetadata = {
      ...attributes.metadata,
      bindings: updatedBindings
    };
    if (customName) {
      updatedMetadata.name = customName;
    }
    setAttributes({
      metadata: updatedMetadata
    });
  }

  // Avoid overwriting other (e.g. meta) bindings.
  if (isConnectedToOtherSources) {
    return null;
  }
  const hasUnsupportedImageAttributes = blockName === 'core/image' && (!!attributes.caption?.length || !!attributes.href?.length);
  const helpText = !hasOverrides && hasUnsupportedImageAttributes ? (0, _i18n.__)(`Overrides currently don't support image captions or links. Remove the caption or link first before enabling overrides.`) : (0, _i18n.__)('Allow changes to this block throughout instances of this pattern.');
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.InspectorControls, {
      group: "advanced",
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.BaseControl, {
        id: controlId,
        label: (0, _i18n.__)('Overrides'),
        help: helpText,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
          __next40pxDefaultSize: true,
          className: "pattern-overrides-control__allow-overrides-button",
          variant: "secondary",
          "aria-haspopup": "dialog",
          onClick: () => {
            if (hasOverrides) {
              setShowDisallowOverridesModal(true);
            } else {
              setShowAllowOverridesModal(true);
            }
          },
          disabled: !hasOverrides && hasUnsupportedImageAttributes,
          accessibleWhenDisabled: true,
          children: hasOverrides ? (0, _i18n.__)('Disable overrides') : (0, _i18n.__)('Enable overrides')
        })
      })
    }), showAllowOverridesModal && /*#__PURE__*/(0, _jsxRuntime.jsx)(_allowOverridesModal.AllowOverridesModal, {
      initialName: attributes.metadata?.name,
      onClose: () => setShowAllowOverridesModal(false),
      onSave: newName => {
        updateBindings(true, newName);
      }
    }), showDisallowOverridesModal && /*#__PURE__*/(0, _jsxRuntime.jsx)(_allowOverridesModal.DisallowOverridesModal, {
      onClose: () => setShowDisallowOverridesModal(false),
      onSave: () => updateBindings(false)
    })]
  });
}
var _default = exports.default = PatternOverridesControls;
